{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cb563d0c",
   "metadata": {},
   "source": [
    "### **Add pre-installed package path & load packages:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c6ea61ad",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(c(.libPaths(), \"/exports/eddie3_apps_local/apps/SL7/R/4.3.0/lib64/R/library\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd87c27e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"dplyr\", warn.conflicts = F)\n",
    "library(\"tidyr\", warn.conflicts = F)\n",
    "library(\"stringr\", warn.conflicts = F)\n",
    "library(\"data.table\", warn.conflicts = F) \n",
    "library(\"vroom\", warn.conflicts = F)\n",
    "library(\"ggplot2\", warn.conflicts = F)\n",
    "#library(survival)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fa61508",
   "metadata": {},
   "source": [
    "### **Change working directory:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b3b6ab1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium'"
      ],
      "text/latex": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium'"
      ],
      "text/markdown": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium'"
      ],
      "text/plain": [
       "[1] \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wd = \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/\"\n",
    "setwd(wd)\n",
    "getwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa16750a",
   "metadata": {},
   "source": [
    "### **Load data:**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0fe9a29",
   "metadata": {},
   "source": [
    "#### Unadjusted proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2ce33c24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>32746</li><li>2951</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 32746\n",
       "\\item 2951\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 32746\n",
       "2. 2951\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 32746  2951"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 2951</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>FID</th><th scope=col>IID</th><th scope=col>Sex</th><th scope=col>Age</th><th scope=col>GeneticArray</th><th scope=col>UKB_centre</th><th scope=col>protein_batch</th><th scope=col>Sample_age</th><th scope=col>PC_1</th><th scope=col>PC_2</th><th scope=col>⋯</th><th scope=col>ZFYVE19</th><th scope=col>ZHX2</th><th scope=col>ZNF174</th><th scope=col>ZNF75D</th><th scope=col>ZNF830</th><th scope=col>ZNRD2</th><th scope=col>ZNRF4</th><th scope=col>ZP3</th><th scope=col>ZP4</th><th scope=col>ZPR1</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1331483</td><td>1331483</td><td>1</td><td>54</td><td>1</td><td>11012</td><td>0</td><td>4612</td><td>-11.1120</td><td>4.32787</td><td>⋯</td><td> 0.4039</td><td>-0.4125</td><td>-0.09345</td><td>-0.76265</td><td>-0.24295</td><td>-0.0419</td><td>-0.07065</td><td>-0.3355</td><td>-0.18575</td><td> 0.48245</td></tr>\n",
       "\t<tr><td>1361634</td><td>1361634</td><td>1</td><td>44</td><td>1</td><td>11010</td><td>0</td><td>5076</td><td>-13.6785</td><td>5.88845</td><td>⋯</td><td> 0.0299</td><td>-0.2598</td><td> 0.18405</td><td>-0.43495</td><td> 3.28595</td><td>-0.2007</td><td> 0.54375</td><td>-0.8688</td><td> 0.28305</td><td> 1.16465</td></tr>\n",
       "\t<tr><td>1508802</td><td>1508802</td><td>0</td><td>54</td><td>1</td><td>11010</td><td>0</td><td>5076</td><td>-12.9177</td><td>6.97354</td><td>⋯</td><td>-0.8821</td><td> 0.3132</td><td> 0.03755</td><td>-0.32695</td><td> 0.25825</td><td>-0.5867</td><td>-0.70445</td><td> 1.1548</td><td>-0.33145</td><td> 1.07485</td></tr>\n",
       "\t<tr><td>1533247</td><td>1533247</td><td>1</td><td>58</td><td>1</td><td>11011</td><td>0</td><td>4982</td><td>-11.5671</td><td>3.51090</td><td>⋯</td><td>-1.6311</td><td> 0.1698</td><td>-0.05835</td><td>-0.32865</td><td> 0.17085</td><td>-0.6857</td><td>-1.00975</td><td>-0.4541</td><td>-0.01685</td><td> 0.21385</td></tr>\n",
       "\t<tr><td>1629003</td><td>1629003</td><td>0</td><td>61</td><td>1</td><td>11016</td><td>0</td><td>4575</td><td>-12.5582</td><td>5.42726</td><td>⋯</td><td>-0.1456</td><td> 0.2485</td><td> 0.85035</td><td> 1.84065</td><td> 0.18675</td><td> 0.0550</td><td>-0.64625</td><td>-0.3087</td><td>-0.38505</td><td>-0.09005</td></tr>\n",
       "\t<tr><td>1629466</td><td>1629466</td><td>0</td><td>58</td><td>1</td><td>11001</td><td>0</td><td>5342</td><td>-13.1117</td><td>2.61686</td><td>⋯</td><td>-0.8135</td><td>-0.1235</td><td>-0.71715</td><td> 0.88495</td><td> 0.41565</td><td>-0.7310</td><td>-0.02125</td><td>-0.8497</td><td>-0.01975</td><td> 2.73195</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 2951\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " FID & IID & Sex & Age & GeneticArray & UKB\\_centre & protein\\_batch & Sample\\_age & PC\\_1 & PC\\_2 & ⋯ & ZFYVE19 & ZHX2 & ZNF174 & ZNF75D & ZNF830 & ZNRD2 & ZNRF4 & ZP3 & ZP4 & ZPR1\\\\\n",
       " <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1331483 & 1331483 & 1 & 54 & 1 & 11012 & 0 & 4612 & -11.1120 & 4.32787 & ⋯ &  0.4039 & -0.4125 & -0.09345 & -0.76265 & -0.24295 & -0.0419 & -0.07065 & -0.3355 & -0.18575 &  0.48245\\\\\n",
       "\t 1361634 & 1361634 & 1 & 44 & 1 & 11010 & 0 & 5076 & -13.6785 & 5.88845 & ⋯ &  0.0299 & -0.2598 &  0.18405 & -0.43495 &  3.28595 & -0.2007 &  0.54375 & -0.8688 &  0.28305 &  1.16465\\\\\n",
       "\t 1508802 & 1508802 & 0 & 54 & 1 & 11010 & 0 & 5076 & -12.9177 & 6.97354 & ⋯ & -0.8821 &  0.3132 &  0.03755 & -0.32695 &  0.25825 & -0.5867 & -0.70445 &  1.1548 & -0.33145 &  1.07485\\\\\n",
       "\t 1533247 & 1533247 & 1 & 58 & 1 & 11011 & 0 & 4982 & -11.5671 & 3.51090 & ⋯ & -1.6311 &  0.1698 & -0.05835 & -0.32865 &  0.17085 & -0.6857 & -1.00975 & -0.4541 & -0.01685 &  0.21385\\\\\n",
       "\t 1629003 & 1629003 & 0 & 61 & 1 & 11016 & 0 & 4575 & -12.5582 & 5.42726 & ⋯ & -0.1456 &  0.2485 &  0.85035 &  1.84065 &  0.18675 &  0.0550 & -0.64625 & -0.3087 & -0.38505 & -0.09005\\\\\n",
       "\t 1629466 & 1629466 & 0 & 58 & 1 & 11001 & 0 & 5342 & -13.1117 & 2.61686 & ⋯ & -0.8135 & -0.1235 & -0.71715 &  0.88495 &  0.41565 & -0.7310 & -0.02125 & -0.8497 & -0.01975 &  2.73195\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 2951\n",
       "\n",
       "| FID &lt;int&gt; | IID &lt;int&gt; | Sex &lt;int&gt; | Age &lt;int&gt; | GeneticArray &lt;int&gt; | UKB_centre &lt;int&gt; | protein_batch &lt;int&gt; | Sample_age &lt;int&gt; | PC_1 &lt;dbl&gt; | PC_2 &lt;dbl&gt; | ⋯ ⋯ | ZFYVE19 &lt;dbl&gt; | ZHX2 &lt;dbl&gt; | ZNF174 &lt;dbl&gt; | ZNF75D &lt;dbl&gt; | ZNF830 &lt;dbl&gt; | ZNRD2 &lt;dbl&gt; | ZNRF4 &lt;dbl&gt; | ZP3 &lt;dbl&gt; | ZP4 &lt;dbl&gt; | ZPR1 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1331483 | 1331483 | 1 | 54 | 1 | 11012 | 0 | 4612 | -11.1120 | 4.32787 | ⋯ |  0.4039 | -0.4125 | -0.09345 | -0.76265 | -0.24295 | -0.0419 | -0.07065 | -0.3355 | -0.18575 |  0.48245 |\n",
       "| 1361634 | 1361634 | 1 | 44 | 1 | 11010 | 0 | 5076 | -13.6785 | 5.88845 | ⋯ |  0.0299 | -0.2598 |  0.18405 | -0.43495 |  3.28595 | -0.2007 |  0.54375 | -0.8688 |  0.28305 |  1.16465 |\n",
       "| 1508802 | 1508802 | 0 | 54 | 1 | 11010 | 0 | 5076 | -12.9177 | 6.97354 | ⋯ | -0.8821 |  0.3132 |  0.03755 | -0.32695 |  0.25825 | -0.5867 | -0.70445 |  1.1548 | -0.33145 |  1.07485 |\n",
       "| 1533247 | 1533247 | 1 | 58 | 1 | 11011 | 0 | 4982 | -11.5671 | 3.51090 | ⋯ | -1.6311 |  0.1698 | -0.05835 | -0.32865 |  0.17085 | -0.6857 | -1.00975 | -0.4541 | -0.01685 |  0.21385 |\n",
       "| 1629003 | 1629003 | 0 | 61 | 1 | 11016 | 0 | 4575 | -12.5582 | 5.42726 | ⋯ | -0.1456 |  0.2485 |  0.85035 |  1.84065 |  0.18675 |  0.0550 | -0.64625 | -0.3087 | -0.38505 | -0.09005 |\n",
       "| 1629466 | 1629466 | 0 | 58 | 1 | 11001 | 0 | 5342 | -13.1117 | 2.61686 | ⋯ | -0.8135 | -0.1235 | -0.71715 |  0.88495 |  0.41565 | -0.7310 | -0.02125 | -0.8497 | -0.01975 |  2.73195 |\n",
       "\n"
      ],
      "text/plain": [
       "  FID     IID     Sex Age GeneticArray UKB_centre protein_batch Sample_age PC_1     PC_2    ⋯ ZFYVE19 ZHX2    ZNF174   ZNF75D   ZNF830   ZNRD2   ZNRF4    ZP3     ZP4      ZPR1    \n",
       "1 1331483 1331483 1   54  1            11012      0             4612       -11.1120 4.32787 ⋯  0.4039 -0.4125 -0.09345 -0.76265 -0.24295 -0.0419 -0.07065 -0.3355 -0.18575  0.48245\n",
       "2 1361634 1361634 1   44  1            11010      0             5076       -13.6785 5.88845 ⋯  0.0299 -0.2598  0.18405 -0.43495  3.28595 -0.2007  0.54375 -0.8688  0.28305  1.16465\n",
       "3 1508802 1508802 0   54  1            11010      0             5076       -12.9177 6.97354 ⋯ -0.8821  0.3132  0.03755 -0.32695  0.25825 -0.5867 -0.70445  1.1548 -0.33145  1.07485\n",
       "4 1533247 1533247 1   58  1            11011      0             4982       -11.5671 3.51090 ⋯ -1.6311  0.1698 -0.05835 -0.32865  0.17085 -0.6857 -1.00975 -0.4541 -0.01685  0.21385\n",
       "5 1629003 1629003 0   61  1            11016      0             4575       -12.5582 5.42726 ⋯ -0.1456  0.2485  0.85035  1.84065  0.18675  0.0550 -0.64625 -0.3087 -0.38505 -0.09005\n",
       "6 1629466 1629466 0   58  1            11001      0             5342       -13.1117 2.61686 ⋯ -0.8135 -0.1235 -0.71715  0.88495  0.41565 -0.7310 -0.02125 -0.8497 -0.01975  2.73195"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# unadjusted proteome file\n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/Resource/Proteome/no_batch7/proteome_unadjusted.txt\"\n",
    "prot <- fread(file) %>% as_tibble\n",
    "# change \"-\" to \"_\" in some column names\n",
    "prot <- rename_with(prot, ~ gsub(\"-\", \"_\", .x))\n",
    "dim(prot)\n",
    "head(prot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ae8ff67",
   "metadata": {},
   "source": [
    "#### Phenotypes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "78396dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "407909"
      ],
      "text/latex": [
       "407909"
      ],
      "text/markdown": [
       "407909"
      ],
      "text/plain": [
       "[1] 407909"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       ".\n",
       "     0      1 \n",
       "400386   7523 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# prelim pheno file\n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/Resource/pheno/delirium.phe\"\n",
    "phe <- fread(file)\n",
    "phe %>% nrow\n",
    "phe %>% pull(delirium) %>% table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "51443fea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# apoe e4 count\n",
    "file = \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/ukb.delirium.gwas/pheno_covar_files/ukb_wb_del_cov_phe_dem_apoe.txt\"\n",
    "apoe <- fread(file, select = c(\"IID\", \"e4_count\"))\n",
    "phe  <- left_join(phe, apoe, by = \"IID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "491d5180",
   "metadata": {},
   "outputs": [],
   "source": [
    "## dementia outcomes\n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/ukb_data/ukb44986/extra_data/dementia_outcomes.csv\"\n",
    "dem  <- fread(file)\n",
    "## select only \"Date of all cause dementia report\"; field 42018\n",
    "dem    <- dem %>% select(eid, `42018-0.0`) %>% filter(`42018-0.0` != \"1900-01-01\") %>% rename(\"dementia_date\" = `42018-0.0`)\n",
    "phe    <- left_join(phe, dem, by = c(\"IID\"=\"eid\")) # merge with main dataset based on \"eid\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "244303cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BMI \n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/ukb_data/ukb44986/extra_data/f21001.BMI.csv\"\n",
    "bmi  <- fread(file) %>% select(eid, `21001-0.0`) %>% rename(\"BMI\" = `21001-0.0`)\n",
    "phe  <- left_join(phe, bmi, by = c(\"IID\"=\"eid\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6c6b959",
   "metadata": {},
   "source": [
    "#### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "a6b0922b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assay table \n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/ukb_data/ukb44986/proteomics/olink_assay.dat\"\n",
    "assay <- fread(file)\n",
    "# replace \"-\" with \"_\" in assay (protein) names\n",
    "assay <- assay %>% mutate(Assay = gsub(\"-\", \"_\", Assay))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "fa3b74cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genomic positions table\n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/Resource/Proteome/Sun_etal_sup/proteins_positions.txt\"\n",
    "pos  <- fread(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "0cac9bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos <-\n",
    "pos %>% \n",
    "    mutate(UniProt = gsub(\";.*\",\"\",UniProt)) %>%\n",
    "    mutate(Gene_CHROM = gsub(\";.*\",\"\",Gene_CHROM)) %>%\n",
    "    mutate(Gene_start = gsub(\";.*\",\"\",Gene_start))%>% \n",
    "    mutate(Gene_end = gsub(\";.*\",\"\",Gene_end)) %>%\n",
    "    filter(!duplicated(UniProt))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82678bd4",
   "metadata": {},
   "source": [
    "### **Exclude withdrawls**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "0108f0e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "w_eids <- fread(\"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/ukb_data/ukb44986/withdraw44986_131_20231013.txt\")\n",
    "w_eids <- w_eids$V1\n",
    "phe    <- phe %>% filter(!(IID %in% w_eids))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a3e1225",
   "metadata": {},
   "source": [
    "### **Make delirium_incident cases outcome, if delirium case reported after 1 year of assessment:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "541f32a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ".\n",
       "    -1      0      1 \n",
       "    75 400386   7448 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>FID</th><th scope=col>IID</th><th scope=col>sex</th><th scope=col>age</th><th scope=col>approx_dob</th><th scope=col>approx_AAA</th><th scope=col>age_death</th><th scope=col>delirium_date</th><th scope=col>censoring_date</th><th scope=col>delirium</th><th scope=col>e4_count</th><th scope=col>dementia_date</th><th scope=col>BMI</th><th scope=col>delirium_incident</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;IDate&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;IDate&gt;</th><th scope=col>&lt;IDate&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;IDate&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1000015</td><td>1000015</td><td>1</td><td>82.0</td><td>1940-11-15</td><td>67.44</td><td>NA</td><td>NA</td><td>2022-10-31</td><td>0</td><td>1</td><td>NA</td><td>24.6832</td><td>0</td></tr>\n",
       "\t<tr><td>1000027</td><td>1000027</td><td>1</td><td>76.1</td><td>1946-09-15</td><td>61.85</td><td>NA</td><td>NA</td><td>2022-10-31</td><td>0</td><td>1</td><td>NA</td><td>32.5577</td><td>0</td></tr>\n",
       "\t<tr><td>1000039</td><td>1000039</td><td>0</td><td>69.0</td><td>1953-10-15</td><td>55.14</td><td>NA</td><td>NA</td><td>2022-10-31</td><td>0</td><td>0</td><td>NA</td><td>19.8903</td><td>0</td></tr>\n",
       "\t<tr><td>1000040</td><td>1000040</td><td>1</td><td>56.6</td><td>1966-03-15</td><td>42.47</td><td>NA</td><td>NA</td><td>2022-10-31</td><td>0</td><td>0</td><td>NA</td><td>30.1242</td><td>0</td></tr>\n",
       "\t<tr><td>1000053</td><td>1000053</td><td>0</td><td>54.7</td><td>1968-02-15</td><td>41.04</td><td>NA</td><td>NA</td><td>2022-10-31</td><td>0</td><td>2</td><td>NA</td><td>22.2860</td><td>0</td></tr>\n",
       "\t<tr><td>1000064</td><td>1000064</td><td>0</td><td>58.6</td><td>1964-03-15</td><td>46.49</td><td>NA</td><td>NA</td><td>2022-10-31</td><td>0</td><td>0</td><td>NA</td><td>44.3343</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " FID & IID & sex & age & approx\\_dob & approx\\_AAA & age\\_death & delirium\\_date & censoring\\_date & delirium & e4\\_count & dementia\\_date & BMI & delirium\\_incident\\\\\n",
       " <int> & <int> & <int> & <dbl> & <IDate> & <dbl> & <dbl> & <IDate> & <IDate> & <int> & <int> & <IDate> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1000015 & 1000015 & 1 & 82.0 & 1940-11-15 & 67.44 & NA & NA & 2022-10-31 & 0 & 1 & NA & 24.6832 & 0\\\\\n",
       "\t 1000027 & 1000027 & 1 & 76.1 & 1946-09-15 & 61.85 & NA & NA & 2022-10-31 & 0 & 1 & NA & 32.5577 & 0\\\\\n",
       "\t 1000039 & 1000039 & 0 & 69.0 & 1953-10-15 & 55.14 & NA & NA & 2022-10-31 & 0 & 0 & NA & 19.8903 & 0\\\\\n",
       "\t 1000040 & 1000040 & 1 & 56.6 & 1966-03-15 & 42.47 & NA & NA & 2022-10-31 & 0 & 0 & NA & 30.1242 & 0\\\\\n",
       "\t 1000053 & 1000053 & 0 & 54.7 & 1968-02-15 & 41.04 & NA & NA & 2022-10-31 & 0 & 2 & NA & 22.2860 & 0\\\\\n",
       "\t 1000064 & 1000064 & 0 & 58.6 & 1964-03-15 & 46.49 & NA & NA & 2022-10-31 & 0 & 0 & NA & 44.3343 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 14\n",
       "\n",
       "| FID &lt;int&gt; | IID &lt;int&gt; | sex &lt;int&gt; | age &lt;dbl&gt; | approx_dob &lt;IDate&gt; | approx_AAA &lt;dbl&gt; | age_death &lt;dbl&gt; | delirium_date &lt;IDate&gt; | censoring_date &lt;IDate&gt; | delirium &lt;int&gt; | e4_count &lt;int&gt; | dementia_date &lt;IDate&gt; | BMI &lt;dbl&gt; | delirium_incident &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1000015 | 1000015 | 1 | 82.0 | 1940-11-15 | 67.44 | NA | NA | 2022-10-31 | 0 | 1 | NA | 24.6832 | 0 |\n",
       "| 1000027 | 1000027 | 1 | 76.1 | 1946-09-15 | 61.85 | NA | NA | 2022-10-31 | 0 | 1 | NA | 32.5577 | 0 |\n",
       "| 1000039 | 1000039 | 0 | 69.0 | 1953-10-15 | 55.14 | NA | NA | 2022-10-31 | 0 | 0 | NA | 19.8903 | 0 |\n",
       "| 1000040 | 1000040 | 1 | 56.6 | 1966-03-15 | 42.47 | NA | NA | 2022-10-31 | 0 | 0 | NA | 30.1242 | 0 |\n",
       "| 1000053 | 1000053 | 0 | 54.7 | 1968-02-15 | 41.04 | NA | NA | 2022-10-31 | 0 | 2 | NA | 22.2860 | 0 |\n",
       "| 1000064 | 1000064 | 0 | 58.6 | 1964-03-15 | 46.49 | NA | NA | 2022-10-31 | 0 | 0 | NA | 44.3343 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  FID     IID     sex age  approx_dob approx_AAA age_death delirium_date censoring_date delirium e4_count dementia_date BMI     delirium_incident\n",
       "1 1000015 1000015 1   82.0 1940-11-15 67.44      NA        <NA>          2022-10-31     0        1        <NA>          24.6832 0                \n",
       "2 1000027 1000027 1   76.1 1946-09-15 61.85      NA        <NA>          2022-10-31     0        1        <NA>          32.5577 0                \n",
       "3 1000039 1000039 0   69.0 1953-10-15 55.14      NA        <NA>          2022-10-31     0        0        <NA>          19.8903 0                \n",
       "4 1000040 1000040 1   56.6 1966-03-15 42.47      NA        <NA>          2022-10-31     0        0        <NA>          30.1242 0                \n",
       "5 1000053 1000053 0   54.7 1968-02-15 41.04      NA        <NA>          2022-10-31     0        2        <NA>          22.2860 0                \n",
       "6 1000064 1000064 0   58.6 1964-03-15 46.49      NA        <NA>          2022-10-31     0        0        <NA>          44.3343 0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# incident delirium: 1\n",
    "# prevalent delirium: -1 (to be excluded)\n",
    "# control: 0\n",
    "\n",
    "phe <- \n",
    "phe %>% \n",
    "    mutate(delirium_incident = case_when((age >= approx_AAA+1 & delirium==1) ~  1, # incident case\n",
    "                                        (age < approx_AAA+1 & delirium==1) ~ -1,  # prevalent case\n",
    "                                        (delirium==0)~0, # control\n",
    "                                         TRUE ~ NA)) \n",
    "phe %>% pull(delirium_incident) %>% table(.,useNA = \"ifany\")\n",
    "head(phe)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca6a1d26",
   "metadata": {},
   "source": [
    "### Make dementia_prevalent covariate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "bab741b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dementia_prevalent</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>407694</td></tr>\n",
       "\t<tr><td>1</td><td>   215</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " dementia\\_prevalent & n\\\\\n",
       " <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 & 407694\\\\\n",
       "\t 1 &    215\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| dementia_prevalent &lt;dbl&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 | 407694 |\n",
       "| 1 |    215 |\n",
       "\n"
      ],
      "text/plain": [
       "  dementia_prevalent n     \n",
       "1 0                  407694\n",
       "2 1                     215"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# dementia_prevalent: \n",
    "# 1 = dementia at assessment + 1 year burn in period after assessment (prevalent); \n",
    "# 0 = after delirium (incident) OR no delirium / dementia\n",
    "\n",
    "phe2 <-\n",
    "phe %>%\n",
    "    # dementia_age variable\n",
    "    mutate(dementia_age = round(as.numeric(difftime(dementia_date, approx_dob, units=\"days\")) / 365.25, digits=1)) %>%\n",
    "    # dementia prior to assessment + 1 year\n",
    "   mutate(dementia_prevalent = case_when(dementia_age <= approx_AAA+1  ~ 1, # prevalent\n",
    "                                         dementia_age  > approx_AAA+1  ~ 0, # incident\n",
    "                                         TRUE ~ 0)) \n",
    "\n",
    "phe2 %>% group_by(dementia_prevalent) %>% summarise(n=n())\n",
    "\n",
    "# dementia prior to delirium (cases) / dementia any time (controls)\n",
    "#    mutate(dementia_prevalent2 = case_when((as.IDate(dementia_date) <= as.IDate(delirium_date)) ~ 1, # prevalent\n",
    "#                                           (as.IDate(dementia_date) >  as.IDate(delirium_date)) ~ 0, # incident\n",
    "#                                            !is.na(dementia_date) ~ 1, # controls with dementia\n",
    "#                                            is.na(dementia_date)  ~ 0, # controls without dementia\n",
    "#                                            TRUE ~ NA)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8851ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#library(lubridate)\n",
    "## adds / substracts dates \n",
    "#phe %>% mutate(a = as.Date(dementia_date) %m-% years(1)) %>% select(dementia_date, a) %>% filter(!is.na(dementia_date))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d24cdeca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>407834</li><li>7</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 407834\n",
       "\\item 7\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 407834\n",
       "2. 7\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 407834      7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>IID</th><th scope=col>sex</th><th scope=col>age_at_collection</th><th scope=col>age</th><th scope=col>BMI</th><th scope=col>e4_count</th><th scope=col>delirium_incident</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1000015</td><td>1</td><td>67.44</td><td>82.0</td><td>24.6832</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><td>1000027</td><td>1</td><td>61.85</td><td>76.1</td><td>32.5577</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><td>1000039</td><td>0</td><td>55.14</td><td>69.0</td><td>19.8903</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1000040</td><td>1</td><td>42.47</td><td>56.6</td><td>30.1242</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1000053</td><td>0</td><td>41.04</td><td>54.7</td><td>22.2860</td><td>2</td><td>0</td></tr>\n",
       "\t<tr><td>1000064</td><td>0</td><td>46.49</td><td>58.6</td><td>44.3343</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " IID & sex & age\\_at\\_collection & age & BMI & e4\\_count & delirium\\_incident\\\\\n",
       " <int> & <int> & <dbl> & <dbl> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1000015 & 1 & 67.44 & 82.0 & 24.6832 & 1 & 0\\\\\n",
       "\t 1000027 & 1 & 61.85 & 76.1 & 32.5577 & 1 & 0\\\\\n",
       "\t 1000039 & 0 & 55.14 & 69.0 & 19.8903 & 0 & 0\\\\\n",
       "\t 1000040 & 1 & 42.47 & 56.6 & 30.1242 & 0 & 0\\\\\n",
       "\t 1000053 & 0 & 41.04 & 54.7 & 22.2860 & 2 & 0\\\\\n",
       "\t 1000064 & 0 & 46.49 & 58.6 & 44.3343 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 7\n",
       "\n",
       "| IID &lt;int&gt; | sex &lt;int&gt; | age_at_collection &lt;dbl&gt; | age &lt;dbl&gt; | BMI &lt;dbl&gt; | e4_count &lt;int&gt; | delirium_incident &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1000015 | 1 | 67.44 | 82.0 | 24.6832 | 1 | 0 |\n",
       "| 1000027 | 1 | 61.85 | 76.1 | 32.5577 | 1 | 0 |\n",
       "| 1000039 | 0 | 55.14 | 69.0 | 19.8903 | 0 | 0 |\n",
       "| 1000040 | 1 | 42.47 | 56.6 | 30.1242 | 0 | 0 |\n",
       "| 1000053 | 0 | 41.04 | 54.7 | 22.2860 | 2 | 0 |\n",
       "| 1000064 | 0 | 46.49 | 58.6 | 44.3343 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  IID     sex age_at_collection age  BMI     e4_count delirium_incident\n",
       "1 1000015 1   67.44             82.0 24.6832 1        0                \n",
       "2 1000027 1   61.85             76.1 32.5577 1        0                \n",
       "3 1000039 0   55.14             69.0 19.8903 0        0                \n",
       "4 1000040 1   42.47             56.6 30.1242 0        0                \n",
       "5 1000053 0   41.04             54.7 22.2860 2        0                \n",
       "6 1000064 0   46.49             58.6 44.3343 0        0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# clean pheno table\n",
    "phe2 <-\n",
    "#phe2 %>% \n",
    "phe %>%\n",
    "    filter(delirium_incident %in% c(0,1)) %>% # exclude prevalent delirium\n",
    "#   mutate(dementia_prevalent = as.factor(dementia_prevalent)) %>% \n",
    "    rename(age_at_collection = \"approx_AAA\") %>%\n",
    "#   select(IID,sex,age_at_collection,age,BMI,dementia_prevalent,delirium_incident)\n",
    "    select(IID,sex,age_at_collection,age,BMI,e4_count,delirium_incident)\n",
    "dim(phe2)\n",
    "head(phe2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10bfdf3d",
   "metadata": {},
   "source": [
    "### Statistics on missingness:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "4590426a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22m`funs()` was deprecated in dplyr 0.8.0.\n",
      "\u001b[36mℹ\u001b[39m Please use a list of either functions or lambdas:\n",
      "\n",
      "# Simple named list: list(mean = mean, median = median)\n",
      "\n",
      "# Auto named with `tibble::lst()`: tibble::lst(mean, median)\n",
      "\n",
      "# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Individuals with > 80% nonmissing values: 32570 out of 32746\"\n",
      "[1] \"Proteins with > 80% nonmissing values: 2919 out of 2923\"\n"
     ]
    }
   ],
   "source": [
    "# statistics on missingness\n",
    "nmiss_id   <- prot[,-c(1:28)] %>% mutate(nmiss_id = rowSums(!is.na(.))/ncol(prot[,-c(1:28)]) ) %>% filter(nmiss_id>0.8) %>% nrow\n",
    "nmiss_prot <- prot[,-c(1:28)] %>% summarise_all(funs(sum(!is.na(.))/nrow(prot[,-c(1:28)]))) %>% t %>% as.vector %>% .[.>0.8] %>% length\n",
    "\n",
    "print(paste(\"Individuals with > 80% nonmissing values:\", nmiss_id, \"out of\", nrow(prot[,-c(1:28)])))\n",
    "print(paste(\"Proteins with > 80% nonmissing values:\", nmiss_prot, \"out of\", ncol(prot[,-c(1:28)])))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bde8f56",
   "metadata": {},
   "source": [
    "### **Merge phe + proteomics data:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "0e886b9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>32746</li><li>2951</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 32746\n",
       "\\item 2951\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 32746\n",
       "2. 2951\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 32746  2951"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>32652</li><li>2954</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 32652\n",
       "\\item 2954\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 32652\n",
       "2. 2954\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 32652  2954"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 2954</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>IID</th><th scope=col>sex</th><th scope=col>age_at_collection</th><th scope=col>age</th><th scope=col>BMI</th><th scope=col>e4_count</th><th scope=col>delirium_incident</th><th scope=col>GeneticArray</th><th scope=col>UKB_centre</th><th scope=col>protein_batch</th><th scope=col>⋯</th><th scope=col>ZFYVE19</th><th scope=col>ZHX2</th><th scope=col>ZNF174</th><th scope=col>ZNF75D</th><th scope=col>ZNF830</th><th scope=col>ZNRD2</th><th scope=col>ZNRF4</th><th scope=col>ZP3</th><th scope=col>ZP4</th><th scope=col>ZPR1</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1000071</td><td>0</td><td>59.17</td><td>74.3</td><td>27.3034</td><td>1</td><td>0</td><td>1</td><td>11004</td><td>3</td><td>⋯</td><td>-0.0404</td><td> 0.0709</td><td> 0.1207</td><td> 0.3909</td><td>-0.2911</td><td>-0.22785</td><td>-0.3313</td><td> 1.6562</td><td> 0.0000</td><td>-0.6093</td></tr>\n",
       "\t<tr><td>1000303</td><td>1</td><td>63.82</td><td>72.1</td><td>30.0092</td><td>0</td><td>0</td><td>0</td><td>11010</td><td>6</td><td>⋯</td><td> 1.8926</td><td>-0.0241</td><td> 0.5815</td><td>-0.1538</td><td>-0.2890</td><td> 0.84655</td><td> 0.1522</td><td> 0.6625</td><td> 0.3080</td><td>-0.2519</td></tr>\n",
       "\t<tr><td>1000629</td><td>0</td><td>54.93</td><td>67.6</td><td>25.6935</td><td>0</td><td>0</td><td>0</td><td>11021</td><td>3</td><td>⋯</td><td>-0.5778</td><td> 0.2080</td><td> 0.1028</td><td>-0.2671</td><td> 0.3807</td><td>-0.38695</td><td> 0.3052</td><td>-4.3959</td><td> 0.1463</td><td> 0.4644</td></tr>\n",
       "\t<tr><td>1000647</td><td>0</td><td>68.51</td><td>76.7</td><td>22.7918</td><td>0</td><td>0</td><td>1</td><td>11010</td><td>3</td><td>⋯</td><td>-0.8161</td><td> 0.0691</td><td>-0.0001</td><td> 0.0606</td><td> 0.6231</td><td>-0.37185</td><td>-0.5297</td><td> 0.8547</td><td>-0.0875</td><td> 1.9491</td></tr>\n",
       "\t<tr><td>1000757</td><td>0</td><td>57.94</td><td>72.2</td><td>25.6568</td><td>1</td><td>0</td><td>1</td><td>11011</td><td>2</td><td>⋯</td><td> 1.0404</td><td>-0.1653</td><td> 0.2392</td><td>-0.1405</td><td> 0.4756</td><td>-0.21525</td><td> 0.8643</td><td> 2.0385</td><td> 0.0294</td><td> 0.0417</td></tr>\n",
       "\t<tr><td>1000900</td><td>0</td><td>47.55</td><td>61.1</td><td>30.9234</td><td>0</td><td>0</td><td>1</td><td>11011</td><td>3</td><td>⋯</td><td> 0.2752</td><td> 0.1067</td><td>-0.1043</td><td>-0.1570</td><td> 0.5793</td><td>-0.18945</td><td>-0.3761</td><td> 0.1647</td><td>-0.2371</td><td>-0.6022</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 2954\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " IID & sex & age\\_at\\_collection & age & BMI & e4\\_count & delirium\\_incident & GeneticArray & UKB\\_centre & protein\\_batch & ⋯ & ZFYVE19 & ZHX2 & ZNF174 & ZNF75D & ZNF830 & ZNRD2 & ZNRF4 & ZP3 & ZP4 & ZPR1\\\\\n",
       " <int> & <int> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <int> & <int> & <int> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1000071 & 0 & 59.17 & 74.3 & 27.3034 & 1 & 0 & 1 & 11004 & 3 & ⋯ & -0.0404 &  0.0709 &  0.1207 &  0.3909 & -0.2911 & -0.22785 & -0.3313 &  1.6562 &  0.0000 & -0.6093\\\\\n",
       "\t 1000303 & 1 & 63.82 & 72.1 & 30.0092 & 0 & 0 & 0 & 11010 & 6 & ⋯ &  1.8926 & -0.0241 &  0.5815 & -0.1538 & -0.2890 &  0.84655 &  0.1522 &  0.6625 &  0.3080 & -0.2519\\\\\n",
       "\t 1000629 & 0 & 54.93 & 67.6 & 25.6935 & 0 & 0 & 0 & 11021 & 3 & ⋯ & -0.5778 &  0.2080 &  0.1028 & -0.2671 &  0.3807 & -0.38695 &  0.3052 & -4.3959 &  0.1463 &  0.4644\\\\\n",
       "\t 1000647 & 0 & 68.51 & 76.7 & 22.7918 & 0 & 0 & 1 & 11010 & 3 & ⋯ & -0.8161 &  0.0691 & -0.0001 &  0.0606 &  0.6231 & -0.37185 & -0.5297 &  0.8547 & -0.0875 &  1.9491\\\\\n",
       "\t 1000757 & 0 & 57.94 & 72.2 & 25.6568 & 1 & 0 & 1 & 11011 & 2 & ⋯ &  1.0404 & -0.1653 &  0.2392 & -0.1405 &  0.4756 & -0.21525 &  0.8643 &  2.0385 &  0.0294 &  0.0417\\\\\n",
       "\t 1000900 & 0 & 47.55 & 61.1 & 30.9234 & 0 & 0 & 1 & 11011 & 3 & ⋯ &  0.2752 &  0.1067 & -0.1043 & -0.1570 &  0.5793 & -0.18945 & -0.3761 &  0.1647 & -0.2371 & -0.6022\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 2954\n",
       "\n",
       "| IID &lt;int&gt; | sex &lt;int&gt; | age_at_collection &lt;dbl&gt; | age &lt;dbl&gt; | BMI &lt;dbl&gt; | e4_count &lt;int&gt; | delirium_incident &lt;dbl&gt; | GeneticArray &lt;int&gt; | UKB_centre &lt;int&gt; | protein_batch &lt;int&gt; | ⋯ ⋯ | ZFYVE19 &lt;dbl&gt; | ZHX2 &lt;dbl&gt; | ZNF174 &lt;dbl&gt; | ZNF75D &lt;dbl&gt; | ZNF830 &lt;dbl&gt; | ZNRD2 &lt;dbl&gt; | ZNRF4 &lt;dbl&gt; | ZP3 &lt;dbl&gt; | ZP4 &lt;dbl&gt; | ZPR1 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1000071 | 0 | 59.17 | 74.3 | 27.3034 | 1 | 0 | 1 | 11004 | 3 | ⋯ | -0.0404 |  0.0709 |  0.1207 |  0.3909 | -0.2911 | -0.22785 | -0.3313 |  1.6562 |  0.0000 | -0.6093 |\n",
       "| 1000303 | 1 | 63.82 | 72.1 | 30.0092 | 0 | 0 | 0 | 11010 | 6 | ⋯ |  1.8926 | -0.0241 |  0.5815 | -0.1538 | -0.2890 |  0.84655 |  0.1522 |  0.6625 |  0.3080 | -0.2519 |\n",
       "| 1000629 | 0 | 54.93 | 67.6 | 25.6935 | 0 | 0 | 0 | 11021 | 3 | ⋯ | -0.5778 |  0.2080 |  0.1028 | -0.2671 |  0.3807 | -0.38695 |  0.3052 | -4.3959 |  0.1463 |  0.4644 |\n",
       "| 1000647 | 0 | 68.51 | 76.7 | 22.7918 | 0 | 0 | 1 | 11010 | 3 | ⋯ | -0.8161 |  0.0691 | -0.0001 |  0.0606 |  0.6231 | -0.37185 | -0.5297 |  0.8547 | -0.0875 |  1.9491 |\n",
       "| 1000757 | 0 | 57.94 | 72.2 | 25.6568 | 1 | 0 | 1 | 11011 | 2 | ⋯ |  1.0404 | -0.1653 |  0.2392 | -0.1405 |  0.4756 | -0.21525 |  0.8643 |  2.0385 |  0.0294 |  0.0417 |\n",
       "| 1000900 | 0 | 47.55 | 61.1 | 30.9234 | 0 | 0 | 1 | 11011 | 3 | ⋯ |  0.2752 |  0.1067 | -0.1043 | -0.1570 |  0.5793 | -0.18945 | -0.3761 |  0.1647 | -0.2371 | -0.6022 |\n",
       "\n"
      ],
      "text/plain": [
       "  IID     sex age_at_collection age  BMI     e4_count delirium_incident GeneticArray UKB_centre protein_batch ⋯ ZFYVE19 ZHX2    ZNF174  ZNF75D  ZNF830  ZNRD2    ZNRF4   ZP3     ZP4     ZPR1   \n",
       "1 1000071 0   59.17             74.3 27.3034 1        0                 1            11004      3             ⋯ -0.0404  0.0709  0.1207  0.3909 -0.2911 -0.22785 -0.3313  1.6562  0.0000 -0.6093\n",
       "2 1000303 1   63.82             72.1 30.0092 0        0                 0            11010      6             ⋯  1.8926 -0.0241  0.5815 -0.1538 -0.2890  0.84655  0.1522  0.6625  0.3080 -0.2519\n",
       "3 1000629 0   54.93             67.6 25.6935 0        0                 0            11021      3             ⋯ -0.5778  0.2080  0.1028 -0.2671  0.3807 -0.38695  0.3052 -4.3959  0.1463  0.4644\n",
       "4 1000647 0   68.51             76.7 22.7918 0        0                 1            11010      3             ⋯ -0.8161  0.0691 -0.0001  0.0606  0.6231 -0.37185 -0.5297  0.8547 -0.0875  1.9491\n",
       "5 1000757 0   57.94             72.2 25.6568 1        0                 1            11011      2             ⋯  1.0404 -0.1653  0.2392 -0.1405  0.4756 -0.21525  0.8643  2.0385  0.0294  0.0417\n",
       "6 1000900 0   47.55             61.1 30.9234 0        0                 1            11011      3             ⋯  0.2752  0.1067 -0.1043 -0.1570  0.5793 -0.18945 -0.3761  0.1647 -0.2371 -0.6022"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data <- inner_join(x=phe2, y=prot, by=\"IID\") %>% select(-c('Age', 'Sex', 'FID')) \n",
    "dim(prot)\n",
    "dim(data) # some missing from prot table: due to extra qc steps in extractF05.ipynb\n",
    "head(data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2be6c27",
   "metadata": {},
   "source": [
    "### **QC proteins with high missingness (>20%)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ba8cacc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "proteins with <80% non-missing rate to be excluded: CTSS GLIPR1 NPM1 PCOLCE"
     ]
    }
   ],
   "source": [
    "n_prot     <- nrow(data[,-c(1:31)]) # #of proteins\n",
    "prot_names <- names(data)[-c(1:31)] # protein names\n",
    "\n",
    "## find proteins with < 80% nonmissing values\n",
    "\n",
    "# produce table with % non-missing per protein\n",
    "nmiss_prot <- data %>% select(all_of(prot_names)) %>% summarise(across(everything(), ~ sum(!is.na(.)) / n_prot)) \n",
    "# protein names with < 80% nonmissing values\n",
    "low_nmiss_prot_names <-  names(nmiss_prot)[nmiss_prot <  0.8] \n",
    "cat(\"proteins with <80% non-missing rate to be excluded:\", low_nmiss_prot_names)\n",
    "\n",
    "# exlcude proteins with <80% non-missing rate\n",
    "data_qc <- data %>% select(-c(all_of(low_nmiss_prot_names)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce61a6b4",
   "metadata": {},
   "source": [
    "### **Statistics on delirium phenotype:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "3a191ff1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dementia_prevalent</th><th scope=col>n</th><th scope=col>percent</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>32638</td><td> 99.957</td></tr>\n",
       "\t<tr><td>1</td><td>   14</td><td>  0.043</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " dementia\\_prevalent & n & percent\\\\\n",
       " <fct> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0 & 32638 &  99.957\\\\\n",
       "\t 1 &    14 &   0.043\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 3\n",
       "\n",
       "| dementia_prevalent &lt;fct&gt; | n &lt;int&gt; | percent &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 0 | 32638 |  99.957 |\n",
       "| 1 |    14 |   0.043 |\n",
       "\n"
      ],
      "text/plain": [
       "  dementia_prevalent n     percent\n",
       "1 0                  32638  99.957\n",
       "2 1                     14   0.043"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## overall cases count\n",
    "data_qc %>% \n",
    "    group_by(dementia_prevalent) %>% \n",
    "#   group_by(delirium_incident, dementia_prevalent) \n",
    "    summarise(n=n()) %>% \n",
    "    mutate(percent = format(n/sum(n)*100,digits = 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c51a088c-8771-4847-8244-1259bede0cf5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>delirium_incident</th><th scope=col>n</th><th scope=col>percent</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>32111</td><td>98.3</td></tr>\n",
       "\t<tr><td>1</td><td>  541</td><td> 1.7</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " delirium\\_incident & n & percent\\\\\n",
       " <dbl> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0 & 32111 & 98.3\\\\\n",
       "\t 1 &   541 &  1.7\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 3\n",
       "\n",
       "| delirium_incident &lt;dbl&gt; | n &lt;int&gt; | percent &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 0 | 32111 | 98.3 |\n",
       "| 1 |   541 |  1.7 |\n",
       "\n"
      ],
      "text/plain": [
       "  delirium_incident n     percent\n",
       "1 0                 32111 98.3   \n",
       "2 1                   541  1.7   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## overall cases count\n",
    "data_qc %>% \n",
    "    group_by(delirium_incident) %>% \n",
    "#   group_by(delirium_incident, dementia_prevalent) \n",
    "    summarise(n=n()) %>% \n",
    "    mutate(percent = format(n/sum(n)*100,digits = 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a6c97c3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2921</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>delirium_incident</th><th scope=col>PC_20</th><th scope=col>A1BG</th><th scope=col>AAMDC</th><th scope=col>AARSD1</th><th scope=col>ABCA2</th><th scope=col>ABHD14B</th><th scope=col>ABL1</th><th scope=col>ABO</th><th scope=col>ABRAXAS2</th><th scope=col>⋯</th><th scope=col>ZFYVE19</th><th scope=col>ZHX2</th><th scope=col>ZNF174</th><th scope=col>ZNF75D</th><th scope=col>ZNF830</th><th scope=col>ZNRD2</th><th scope=col>ZNRF4</th><th scope=col>ZP3</th><th scope=col>ZP4</th><th scope=col>ZPR1</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>32111</td><td>32111</td><td>32111</td><td>31945</td><td>32111</td><td>31831</td><td>31945</td><td>32111</td><td>32111</td><td>⋯</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td><td>32111</td></tr>\n",
       "\t<tr><td>1</td><td>  541</td><td>  541</td><td>  541</td><td>  533</td><td>  541</td><td>  531</td><td>  533</td><td>  541</td><td>  541</td><td>⋯</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td><td>  541</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2921\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " delirium\\_incident & PC\\_20 & A1BG & AAMDC & AARSD1 & ABCA2 & ABHD14B & ABL1 & ABO & ABRAXAS2 & ⋯ & ZFYVE19 & ZHX2 & ZNF174 & ZNF75D & ZNF830 & ZNRD2 & ZNRF4 & ZP3 & ZP4 & ZPR1\\\\\n",
       " <dbl> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & ⋯ & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 & 32111 & 32111 & 32111 & 31945 & 32111 & 31831 & 31945 & 32111 & 32111 & ⋯ & 32111 & 32111 & 32111 & 32111 & 32111 & 32111 & 32111 & 32111 & 32111 & 32111\\\\\n",
       "\t 1 &   541 &   541 &   541 &   533 &   541 &   531 &   533 &   541 &   541 & ⋯ &   541 &   541 &   541 &   541 &   541 &   541 &   541 &   541 &   541 &   541\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2921\n",
       "\n",
       "| delirium_incident &lt;dbl&gt; | PC_20 &lt;int&gt; | A1BG &lt;int&gt; | AAMDC &lt;int&gt; | AARSD1 &lt;int&gt; | ABCA2 &lt;int&gt; | ABHD14B &lt;int&gt; | ABL1 &lt;int&gt; | ABO &lt;int&gt; | ABRAXAS2 &lt;int&gt; | ⋯ ⋯ | ZFYVE19 &lt;int&gt; | ZHX2 &lt;int&gt; | ZNF174 &lt;int&gt; | ZNF75D &lt;int&gt; | ZNF830 &lt;int&gt; | ZNRD2 &lt;int&gt; | ZNRF4 &lt;int&gt; | ZP3 &lt;int&gt; | ZP4 &lt;int&gt; | ZPR1 &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0 | 32111 | 32111 | 32111 | 31945 | 32111 | 31831 | 31945 | 32111 | 32111 | ⋯ | 32111 | 32111 | 32111 | 32111 | 32111 | 32111 | 32111 | 32111 | 32111 | 32111 |\n",
       "| 1 |   541 |   541 |   541 |   533 |   541 |   531 |   533 |   541 |   541 | ⋯ |   541 |   541 |   541 |   541 |   541 |   541 |   541 |   541 |   541 |   541 |\n",
       "\n"
      ],
      "text/plain": [
       "  delirium_incident PC_20 A1BG  AAMDC AARSD1 ABCA2 ABHD14B ABL1  ABO   ABRAXAS2 ⋯ ZFYVE19 ZHX2  ZNF174 ZNF75D ZNF830 ZNRD2 ZNRF4 ZP3   ZP4   ZPR1 \n",
       "1 0                 32111 32111 32111 31945  32111 31831   31945 32111 32111    ⋯ 32111   32111 32111  32111  32111  32111 32111 32111 32111 32111\n",
       "2 1                   541   541   541   533    541   531     533   541   541    ⋯   541     541   541    541    541    541   541   541   541   541"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"No. of delirium cases summary:\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "  466.0   532.0   533.0   535.5   541.0   541.0 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAYAAAD958/bAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdebhtV13n608a0hLAUEHSIITOomjEGEKjXjUEhMKiF+pWUGwgVUoBVpUX\nYomoiJZ61QIuhSK9oiIQCKCAApFGkCIQMIA0IkQsAgQIoUlHmnP/mGM/2Wdnn3P2PrtZK2u/\n7/PMZ+y15lhr/daCh+TLmPM3CgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGA3B8y6AAA2xduqg5Y9/tnq\nH2ZUCzdMd6/+ffWvq5tUB47nz6meMauiALbbwbMuAGCHen71ncsev696yh7mnlk9YNnjL1U/\numLOD7R7QLrJRgsc7lgdt+zxhdUnN+m9mR//T/Vbrf5/nH5hm2sBAGAHen+1a9nx5r3MfemK\nuf9nlTlXr5hzr02q8w9WvO8fbNL7Mj9u3fX/+7P8eMXsSgPYflaQABbDh9r9f9MvnVUh3OD8\nULuvPn6temT1T00ByX+XgB1FQAJYDCfPugBusI5e8fiD1VtnUQjAPBCQABbDWpo03Kl6WHXn\n6uZNqwNfbrrH5ENNl/19cjz/zOr7mu5BWu7BTTfxL/la9ZAVcw6rHlp9f3VCdXh1cfX31Ruq\nj+zju/xQU7OAW1dfbWoS8PJx7k0r5v5k9Zllj1f7HT7W9L3/3ajno9XPjfPHVidV3119R1NY\nuGl1edPv8oHqrKbfaTVvX/H4Z6tPVT9e3a86svpw9fvVZ5fNO7X6D6OeLzX9Lq9q+u3313p/\n95+pHl3dasXzd2/37/XL1Ts2UBcAAOzTdt+D9LvVte35PpOl47Qx/+w1zF0KWMs9silY7O01\nr6putsp3OLB64R5e87Hqbqs8f5d9/A4/WP3liufevmz+h9bwHa/oukC10sq5D67OW+X5b1T3\nbQpvL97D57ym6zrHrdf+/O6/s4bvvmu8NwAAbKmVAekD1Y/s4XjrirnrDUg/2tr+RXijAekn\n1/E5H2paXVnut/bxms+u8ty+AtLK33lXu6+GrCUgLR3/vutbOWdvIeWLTStJe/uMM1b5jH3Z\n399dQAIAYG6s9i/uaz3WG5D+bMW5Z1X3bgoXP9B0Wdgrm1Y5lgLSqdVPVH+z4rV/M55fOpZC\nwwlNKy3L536l+s2m1Zc3rfI9/ueyGu+wynf4avV71S/u5ffaV0BaHuTe1xSy3rls/vuaLj/7\nj9W/bbrU7uTqUdXfrniPj3d9q33Wv1TPW+X1S8e1TatFL64uW3Hufat8xt5s5Hc/uek/w1es\nOP/Rdv/P+NbrrAkAANZtOwPSXy97/pqme1NWc2TT/TfLrbXN92+smPetdr9XqeplK+Zc1nWr\nGb++4txV7R5+Dq3e3fV/i7UEpF9o9/uSvmPZ34ft4ftUHTXqWP5eK+/XWflZFzbd31XT5XIf\nW2XOk5a9/j92/d/tRnupaaWN/u5VP7/ivAYNwI62v9c6A3DD8S/L/j6wek/1a9XpTQ0KlkLC\npU1NF/bHaSsev6brr7j8jxWPD6/uM/7+/hXnXt/uTQWubLoEb73eOT73mmXPLW+WcEXTKszP\nNm3e+4amRg9vH3+vdNt9fN6LmlZwalopev+K81+r/nDZ43NWnL9Rq9+ftScb/d0BWEEXO4D5\ncG71lD2cO7P64Q2898ua7lM5YDy++ziWfKspSDyrqaHB/rjNiscfXWXOJ5tWeJb/s2fpdStX\nZj68yutXe25fXr6P889o+n3XumqzcoVtpQ+seHzJiscfaQplS76+ynscusZaauO/OwArCEgA\n8+Hirt8yeslPbPC931n9WFMnu29f5fwhTSsRp1X/ufpf+/EZKy9V+8Yqc65turzrJqu8buXr\nv7nK61d7bl8u2Mu5x1W/tOK5q6vPN9VZdft2vzxvX1defGnF42tWPP78iscb/efwRn93AFZw\niR3AzvAnTas0pzU1Pfij6u+aLl1b7lfbv382fGXF4+NXmXNku/9L+vLXXbzi+eNWef0J+1HX\naoFhycqOcW9tqvs7mu7juVO7r/asxcpAtK/zB6w6a+02+rsDsIKABLBzXNV0f81vVI9tug/l\nxHa/DOzm7R5EVv4L/Z5WPFbea/Pgrv8v/w/by+vOX/H8g1Z5/Wa3m165Ce7zqouWPb5L129F\nPm82+rtvph9qulxx6XjoFnwGwJYTkAAW33+untzqqwu72v0SspouM1uysmnDPaojVnmfP1/x\n+I7V/9d1AeOeXb/Jwoea7o+pet0eXn+zpvuDHtWeN2vdXytXh75v2d9Ht+eOffNko7/7ZnpQ\nU0OIpeMxW/AZAAAsqJVtvt+8l7kvXTF3vW2+n7Xs+U83tf1+9Ri/seJ1/9zuKxD/d9dvU335\nmHdB9fQx74BW3/fn6qb9jFY+f01132Wfc6PqH1aZd23TZYB7anm+rzbf92rPXrnK+72/emPT\nJWirfd7KVZGV51d+3rNWnH/FivMnrPIe67mUcKO/e21em++VG8++ej/fB2CmrCAB7CwnVver\nHjHGGy87d03TStOuZc/9RdMmq8sd1nSfzq2bVloar3l41+/idlDXb1t9VdP+P29b8dzpXX/F\n6oCmJhI13R+10r7u+dmbX21qbb7c91QPbPpez2/+79XZ6O8OwAoCEsDie11TU4Z/3MP5q6q/\nqn6gOnvFuW80XTq1lhbbF1XfW/236hOrnL+sOqvpsq8XrnL+g02Xub1rxfOfrf5D9WervGZl\nc4f1+Gh1atMlZ8t9uXpq095INwQb/d0BWGaj3XMA2D/fUx217PHFXb9RwZJ/Xd1y2eMrmzrQ\nLfcD7f6/6ee1+h47R1a3qI5parjw1eqfmvZC2pcTmjrhLd+n5/9Un9rD/G8frzl0fM6nmsLY\nWix1k7u4KdhdWz2+3TdZ/dL4jOUrXmv9HVa6TVPnvEu6bt+gmoLH8j2SPtLuK2o/uOJ9Vn7e\n7dv9krmLmi4lXHJode8V77Fad8H1WO/vfqvqdssef7X6+/343Nu1+35WX2r1fZkAAIA1uEfT\nnk83XuXcSU3BZPk9Ls/ftsoAAAC22UO7rgnE31WvGscHu36zga+2+2oFAADAQlkKSPs6vtC0\nhxMAbLqVe18AwKzsarp/5hatvkHrPzVt5vrjbc0+PgCgSQMAc+eApqB0THWT6ptNzSDmveU2\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAsEkOmHUBAADAqm5SnbIJ7/PR6vOb8D47wsGzLgAAAFjV\nGQcddND/e+SRR+73G1x++eVdddVVL6oet3llLTYBCQAA5tPBd77znfuTP/mT/X6Dpz3tab3u\nda87aBNrWngHzroAAACAeSEgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAA\nDAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEA\nAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICAB\nAAAMAhIAAMBw8KwL2GY3re5WHVsdUV1Wfb46v/raDOsCAADmwE4JSKdUz6hOqw5a5fzV1Vur\np1fnbmNdAADAHNkJAenR1cubvuuXqw80rRpdUR3WtJp0cvWApgB1evXKmVQKAADM1KIHpKOr\nF1SXVGdUZ1e7Vpl3YPWw6vnVC5tWky7ephoBAIA5sehNGh5UHVX9dPXaVg9HVddWZ1WPG/Mf\nuC3VAQAAc2XRA9IJY3zHGue/fYy32vxSAACAebfoAWmpM92Ja5x/2zFesgW1AAAAc27RA9Lb\nmi6re151zD7m3mLM21Wds8V1AQAAc2jRmzR8oin0PKH6dPW6pjbeF1ZXVodWx1f3qB5SHVk9\nt/rkLIoFAABma9EDUtWTqouqM5taeJ++h3mXN+2D9MxtqgsAAJgzOyEgXdu0Seyzm7rTnVQd\nVx3eFIourM6r3lh9fUY1AgAAc2AnBKQlX6teMQ4AAIDrWfQmDQAAAGu2k1aQqm5a3a06tjqi\nuqz6fHV+17UEBwAAdqidEpBOaboP6bTqoFXOX129talJw7nbWBcAADBHdkJAenT18qbv+uXq\nA02rRldUhzWtJp1cPaApQJ1evXImlQIAADO16AHp6OoF1SXVGdXZTRvBrnRg9bDq+dULm1aT\nLt6mGgEAgDmx6E0aHlQdVf109dpWD0c1tQI/q3rcmP/AbakOAACYK4sekE4Y4zvWOP/tY7zV\n5pcCAADMu0UPSEud6U5c4/zbjvGSLagFAACYc4sekN7WdFnd86pj9jH3FmPeruqcLa4LAACY\nQ4vepOETTaHnCdWnq9c1tfG+sLqyOrQ6vrpH9ZDqyOq51SdnUSwAADBbix6Qqp5UXVSd2dTC\n+/Q9zLu8aR+kZ25TXQAAwJzZCQHp2qZNYp/d1J3upOq46vCmUHRhdV71xurrM6oRAACYAzsh\nIC35WvWKcWy1o6v/2RTC1uKQpo57J29ZRQAAwD7tpIC0nXZV36iuWOP8b6++pykofWurigIA\nAPZOQNoaX63+8zrm36epSQQAADBDi97me8ltm+5DelH189XN9zDv26s3V4/ZproAAIA5shNW\nkH6o+st2vx/ov1ePrd6wYu7h1Q9X792e0gAAgHmy6CtIN6pe1hR8Xl09vvq96rDq7KaQBAAA\nUC3+CtIPVbeq/rj68WXP/2H1F9WLq6uqP93+0gAAgHmz6CtIdxnji1c8/4nqe6uPV39UPXI7\niwIAAObTogekQ8Z40SrnLqpOqz7TtIL0I9tVFAAAMJ8WPSD9yxhP2MP5z1f3rb7QdI/S/baj\nKAAAYD4tekD60BhP3cuczzatJF1S/f6WVwQAAMytRQ9IH63+sTq9qaPdnnyyafXoa9tRFAAA\nMJ8WvYtd1ROqO1fHVBfuZd6Hq/tXP1r97TbUBQAAzJmdEJDeMo61+MA4AACAHWjRL7EDAABY\nMwEJAABgEJB2d0J1QfVzM64DAACYAQFpdwdXt65uNutCAACA7ScgAQAADIvexe6R41irI7eq\nEAAAYP4tekC6S/XoWRcBAADcMCx6QPrmGH+lte1vdIvqRVtWDQAAMNcWPSCdP8ZvVH+xhvm3\n2bpSAACAebfoTRrOG+N3z7QKAADgBmHRA9KXq79q7d/zkurXq3duWUUAAMDcWvRL7KoesI65\nl1RP26pCAACA+bboK0gAAABrJiABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICAB\nAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAg\nAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAg\nIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAfPuoAZ\nOLg6ujqiuqy6uLp6phUBAABzYaesIB1VnVmdW11afbH6zBi/OZ5/anXjWRUIAADM3k5YQbpr\n9abq+PH4iuqCMR5W3bI6eRxPrB5YfXjbqwQAAGZu0VeQDq1eWx1bPae6W9OldSdWdxrjEdV3\nVc8d815THTKLYgEAgNla9IB0/+p21VOqJzetDO1aMWdXdX7T6tGZ1e3H6wAAgB1m0QPSncb4\nsjXOf+mK1wEAADvIogekpe50R65x/tK8q7agFgAAYM4tekB63xifVh2wj7kHVL80/j53yyoC\nAADm1qJ3sXt39c7qcdVdqhc1hZ8LqyubmjgcX91jzDllzH/PLIoFAABma9ED0q7q4dUrq1Or\ne+1j/jnVo7p+IwcAAGAHWPSAVPWV6r7V/apHVCdVx1WHV5c3rSadV51VvWVGNQIAAHNgJwSk\nJW9JAAIAAPZi0Zs0AAAArNlOWkFacnB1dHVEdVl1cde1AwcAAHawnbKCdFR1ZlMHu0urL1af\nGeM3x/NPrW48qwIBAIDZ2wkrSHet3tTUzrvqiuqCMR5W3bI6eRxPrB5YfXjbqwQAAGZu0VeQ\nDq1eWx1bPae6W9OldSdWdxrjEdV3Vc8d815THTKLYgEAgNla9IB0/+p21VOqJzetDK3c42hX\ndX7T6tGZ1e3H6wAAgB1m0QPSncb4sjXOf+mK1wEAADvIogekpe50R65x/tK8q7agFgAAYM4t\nekB63xifVh2wj7kHVL80/j53yyoCAADm1qJ3sXt39c7qcdVdqhc1hZ8LqyubmjgcX91jzDll\nzH/PLIoFAABma9ED0q7q4dUrq1Ore+1j/jnVo7p+IwcAAGAHWPSAVPWV6r7V/apHVCdVx1WH\nV5c3rSadV51VvWVGNQIAAHNgJwSkJW9p+wLQwdW/a+2/73duYS0AAMAa7aSAtJ2Oq36ztf++\nh41xX40kAACALbQTA9LB1dHVEdVl1cVd1w58s3y29a0K3aepoYR7nwAAYIYWvc33kqOqM5s6\n2F1afbH6zBi/OZ5/anXjWRUIAADM3k5YQbpr9aamdt5VV1QXjPGw6pbVyeN4YvXA6sPbXiUA\nADBzi76CdGj12urY6jnV3ZourTuxutMYj6i+q3rumPea6pBZFAsAAMzWogek+1e3q55SPblp\nZWjlfT67qvObVo/OrG4/XgcAAOwwix6Q7jTGl61x/ktXvA4AANhBFj0gLXWnO3KN85fmXbUF\ntQAAAHNu0QPS+8b4tPa9x9AB1S+Nv8/dsooAAIC5tehd7N5dvbN6XHWX6kVN4efC6sqmJg7H\nV/cYc04Z898zi2IBAIDZWvSAtKt6ePXK6tTqXvuYf071qGzYCgAAO9KiB6Sqr1T3re5XPaI6\nqTquOry6vGk16bzqrOotM6oRAACYAzshIC15SwIQAACwF4vepAEAAGDNBCQAAIBBQNrdv2ra\nLPahM64DAACYAQFpdzeuHlvdfdaFAAAA209AAgAAGBa9i93J41irm29VIQAAwPxb9ID0I9Uv\nz7oIAADghmHRA9JVY3xV9dk1zL9J9fitKwcAAJhnix6QPjbGt1Z/uIb5t0lAAgCAHWvRmzR8\ncIzfPdMqAACAG4RFD0ifqb7a2gPSFdW7W9vleAAAwIJZ9Evsqk5Yx9wvVN+3VYUAAADzbScE\npMtmXQAAAHDDsOiX2AEAAKyZgAQAADAISAAAAIOABAAAMAhIAAAAg4AEAAAwCEgAAACDgAQA\nADAISAAAAIOABAAAMAhIAAAAg4AEAAAwCEgAAACDgAQAADAISAAAAIOABAAAMAhIAAAAg4AE\nAAAwCEgAAACDgAQAADAISAAAAIOABAAAMAhIAAAAg4AEAAAwCEgAAACDgAQAADAISAAAAIOA\nBAAAMAhIAAAAg4AEAAAwCEgAAACDgAQAADAISAAAAIOABAAAMAhIAAAAg4AEAAAwCEgAAACD\ngAQAADAISAAAAIOABAAAMAhIAAAAg4AEAAAwHDzrArbZTau7VcdWR1SXVZ+vzq++NsO6AACA\nObBTAtIp1TOq06qDVjl/dfXW6unVudtYFwAAMEd2QkB6dPXypu/65eoDTatGV1SHNa0mnVw9\noClAnV69ciaVAgAAM7XoAeno6gXVJdUZ1dnVrlXmHVg9rHp+9cKm1aSLt6lGAABgTix6k4YH\nVUdVP129ttXDUdW11VnV48b8B25LdQAAwFxZ9IB0whjfscb5bx/jrTa/FAAAYN4tekBa6kx3\n4hrn33aMl2xBLQAAwJxb9ID0tqbL6p5XHbOPubcY83ZV52xxXQAAwBxa9CYNn2gKPU+oPl29\nrqmN94XVldWh1fHVPaqHVEdWz60+OYtiAQCA2Vr0gFT1pOqi6symFt6n72He5U37ID1zm+oC\nAADmzE4ISNc2bRL77KbudCdVx1WHN4WiC6vzqjdWX59RjQAAwBzYCQFpydeqV4wDAADgeha9\nSQMAAMCa7aQVpKqbVnerjq2OqC6rPl+d33UtwQEAgB1qpwSkU5ruQzqtOmiV81dXb21q0nDu\nNtYFAADMkZ0QkB5dvbzpu365+kDTqtEV1WFNq0knVw9oClCnV6+cSaUAAMBMLXpAOrp6QXVJ\ndUZ1dtNGsCsdWD2sen71wqbVpIu3qUYAAGBOLHpAelB1VPWY6vV7mXdtdVZ1TfXapnbgf7LB\nzz6qtf++R23wswAAgE2w6AHphDG+Y43z3z7GW23wc29X/WN1wDpft975AADAJlr0gLTUme7E\n6kNrmH/bMV6ywc/9p+ouTZvRrsXdqhe3+uV/AADANln0gPS2ptDxvOoh1Zf2MvcWY96u6pxN\n+Ox/WMfcQzfh8wAAgA1a9ID0iabQ84Tq09Xrmtp4X1hd2RRMjq/u0RSgjqyeW31yFsUCAACz\ntegBqepJ1UXVmU0tvE/fw7zLm/ZBeuY21QUAAMyZnRCQrm3aJPbZTd3pTqqOa7o/6PKm1aTz\nqjdWX59RjQAAwBzYCQFpydeqV4wDAADgeg5cx9xv27IqAAAA5sB6AtKTq49UT6xuujXlbLtD\nqu+vHll914xrAQAAZmw9AemfqztUz2m6b+fF1T23oqhNdoum7nVXVhdU/2E8/2+qj1bvrF7V\ntE/Su6tbbn+JAADAPFhPQHpJdUL11KaA9JPVe6sPVv+pOmrTq9scf1Q9uOl+q1uNx/euXtO0\nMezfVK9u+k73yT1KAACwY60nINW00epvV3es7te08nLn6vebAsYfNnWJmxd3qX64ekt183Gc\n07T6dYfq/tWp1Y9W/7p6V/UD1XfPolgAAGC21huQluyq3lo9qmlV5hea9hp6fPWBps1Yf7o6\nbBNq3IjvGeN/ry4Zx39vCkNvqN62bO43xrm6YVw6CAAAbLL9DUjLfbn6WPWPy547uXph9enq\nIZvwGfvr2DEur+1TY/zEKvP/YYw337KKAACAubWRgPQd1a9Wn63Ork5rWpV5YNOq0q9WR1Zn\nVd+5sTL321fGuDzwLP19i1XmLz132ZZVBAAAzK31BqSDqn9X/UX1merp47nfaGp48ODqzdX/\nqX6l6f6eg5ruA5qFpVWixy977vHVNU33H60MST8zxo9tcV0AAMAcOngdc3+kqRnDCePxu6rn\nNa0QXbWH13yg+lZ19P4WuEHvatq76czq/xrP3ad6Y9Pq0vubvtNXmppOPLL6YlNnOwAAYIdZ\nT0A6uWmD2OeN46NrfN2Tq/PWWddm2VX9RPX6pmBU9fGmtuTXNjWT+I1l868c86/ctgoBAIC5\nsZ6A9GfV71TfXOdn/ME652+2DzRd/nfPpuBzXteteN2lOqO6ffW56o+7rokDAACww6wnIK3W\n9e2G4srqnas8f3H1m9tcCwAAMKfW06ThwKbmDL+4lzlPa7qc7YCNFAUAADAL6wlI964eVP39\nXuZ8pKnLnY1WAQCAG5z1BKS7jvH9e5mzdO6ue5kDAAAwl9YTkJZadX9lL3MuGuO/2r9yAAAA\nZmc9AWkpGN1hL3NuP8ZL9q8cAACA2VlPQPrbMT51L3OWzr17/8oBAACYnfUEpI9Wb6p+vPrL\n6nubLrs7evz9l+PcX1Xnb26ZAAAAW289+yBVPbb66+rfjmOlD1U/ttGiAAAAZmG9AelL1b2q\nn6oeWt222lV9ujq7eknTpqwAAAA3OOsNSDUFoN8fBwAAwMJYzz1IAAAAC01AAgAAGNYbkG5c\nPaP6cPWN6qrq6lWOp29ijQAAANtiPfcgHVS9palJwzXV56qv7mHuRRusCwAAYNutJyA9sCkc\nva2p3ffntqQiAACAGVnPJXZ3GeOTE44AAIAFtJ6AdOkYXT4HAAAspPUEpLeN8Z5bUQgAAMCs\nrScg/UP1W9WzqjtuTTkAAACzs54mDQ+tbll9W/WR6r3VPzd1tFvp7HEAAADcYKwnIN29qXvd\nku8fx2ouSEACAABuYNYTkP6gtYeeL+xHLQAAADO1noD0hQQfAABgga0nIC13ePWdTfcffXjz\nygEAAJid9XSxq7pR9TvVxdUHq99fdu6g6pzqzzenNAAAgO213oD0/Oq/VR9ravu93DXVu6pH\nVrfYeGkAAADbaz0B6Y7VT1R/Wp1c/c0qc14/3vM+G64MAABgm60nIJ1aHVD9WnXtHuZ8fIwn\nbKQoAACAWVhPQLr5GC/Yy5ylTWOP2q9qAAAAZmg9AemrYzxmL3O+e4yf379yAAAAZmc9Aend\nY3zEHs4f3HT53a7qnRspCgAAYBbWsw/S31d/Vf1602rSEeP5W1X3rH6lunNTm+9Pb16JAAAA\n22O9G8X+ePWm6qXLnvvssr/fV/2nDdYEAAAwE+sNSBdV964eWz20un3TZXqfqV5Tvbj61mYW\nCAAAsF3WG5BqCkAvGAcAAMDCWE+TBgAAgIUmIAEAAAzrucTuSeNYi+eMAwAA4AZjPQHpiOpf\nrfL8QdWNx9/fqi7vuhbgAAAANxjrucTuN6ubrXIcVR1TPbX6RvWjYy4AAMANyv50sVvNl6vf\nrg5o2ij2ttUlm/TeAAAA22KzmzS8qvq26sGb/L4AAABbbrMD0pVjPH6T3xcAAGDLbXZAevwY\nP7fJ7wsAALDl1nMP0g+OYzU3q06p7l19pXrDhqoCAACYgfUGpF/ex5zzq5+qvrq/BQEAAMzK\negLSn1bv38O5K6vPVJ/acEUAAAAzsp6A9MlxAAAALKTN2gfphuTg6ujqiOqy6sVF7poAACAA\nSURBVOLq6plWBAAAzIXN7mI3r46qzqzOrS6tvth0SeAXq2+O559a3XhWBQIAALO3nhWknxvH\n/njWOGbhrtWbum5vpiuqC8Z4WHXL6uRxPLF6YPXhba8SAOCG73bVSRt8j2urt1Zf23g5sH7r\nCUiHNK2wHDX+rvrWsnNLj7+xh9fOwqHVa6tjq+dUL6w+Uu1aNueAphD1+Opnq9dUd+667wYA\nwNr8xiGHHPKoww47bL/f4Jvf/GbXXnvtz1R/sHllwdqtJyD9dvXP1Quq36ueX312nPuO6j9W\nTxjHn29ijRtx/6b/J+Pnq9/dw5xdTe3Jn9j0fX57vO4vtqNAAIAFctAjH/nIfuEXfmG/3+Bh\nD3tYn/rUpw7axJpgXdZzD9Itq5dUT6l+oekytWvHccF47iljzi03s8gNuNMYX7bG+S9d8ToA\nAGAHWU9A+pHq8OqP9zLnj8ecB22kqE201J3uyDXOX5p31RbUAgAAzLn1BKRjxnjAXuYsnTtm\nL3O20/vG+LT2Xnfj/C+Nv8/dsooAAIC5tZ6AdMEYH7uXOT8xxs/sTzFb4N3VO6vHVe8Z43c1\nBbibjPHuTQ0a3lv91Jj/nlkUCwAAzNZ6mjS8obqwqV33idUfdl0QOrE6o3py9bnmp8HBrurh\n1SurU6t77WP+OdWj2r3LHQAAsEOsJyB9s3pIU1D6b+O4dpxbWon6wphz6WYVuAm+Ut23ul/1\niKbe/Mc13St1eVPoO686q3rLjGoEAADmwHoCUtX7q3/TtFr0w03tvWtqj/1XTatKX9206jbX\nWxKAAACAvVhvQKopAP3WOAAAABbG/gSkmi5P+87qmurDm1fOtji4Oro6orqsurjr2oEDAAA7\n2Hq62FXdqPqdplDxwer3l507qKnJwZ9vTmmb6qjqzKb23ZdWX2xqMPHFpnurzq2eWt14VgUC\nAACzt94VpOdXP9kUjg5dce6a6l1New7dorpow9VtjrtWb6qOH4+vaGpZfkV1WHXL6uRxPLF6\nYDe8VTEAAGATrGcF6Y5N+xz9aVOY+JtV5rx+vOd9NlzZ5ji0em11bPWc6m5Nl9adWN1pjEc0\n7Y303DHvNdUhsygWAACYrfUEpFOrA6pf67r23it9fIwnbKSoTXT/6nbVU5r2aPpw19/jaFd1\nftPq0ZnV7cfrAACAHWY9l9jdfIwX7GXONWM8ar+q2Xx3GuPL1jj/pdVvj9dtZLPb46pXt/aV\nqKV7nw7YwGcCAAAbtJ6AtLS/0THVv+xhzneP8fP7XdHmWupOd2T15TXMP3KMV23wcy9ualZx\n2Brn37qpK+DK1S0AAGAbrScgvXuMj6ietYf3+rWmf8l/5wbr2izvG+PTmja33VsAOaD6pfH3\nuRv83CuqZ69j/n2qn9ngZwIAABu0noD099VfVb/etJp0xHj+VtU9q1+p7ty0cvLpzStxQ97d\nFNYeV92lelFT+LmwurKpicPx1T3GnFPG/PfMolgAAGC21tvm+8ebWma/dNlzn1329/uq/7TB\nmjbTrurh1Submkzcax/zz6kelUvdAABgR1pvQLqounf12OqhTR3fDmzadPU11Yurb21mgZvg\nK9V9q/s1XR54UlMThcOry5tWk86rzqreMqMaAQCAObCegHTCOD5VvWAcNyRvSQACAAD2Yj37\nIP1U9XfVLbeoFgAAgJlazwrSV8Z42VYUMiPHNK2KfbNpZcy9RwAAsIOtZwXpHWO891YUsoV+\nqLrbiufuXv1t0z1V51WfrL5QPWF7SwMAAObJegLSR6rnVL/b1Bb7huC2TZ3pHrrsuTs1tfL+\n3uob1YeaOvEdUz236/ZCAgAAdpj1XGL30Oqm1Y2a2nmf17Tf0aWrzD17HLP2fWP822XP/Vp1\nVPUb1TOa9kOqaQ+kVzVtKvuipu52AABso8997nM1/Z/Wz93A21xdndy0jyesy3oC0t2b2nsv\nOWkcq7mg+QhIx47xC8ueu3/1ruoXV8x9X9Mldm8Yc1661cUBALC7a665ptNPP70f/MEf3O/3\nOOOMMw7etWvXzTevKnaS9QSkP2jtoecL+56yLZZWh44c4yFNq0fv28P8947xuK0sCgCAPbvN\nbW7Tve51r/1+/QEHHNCuXXpvsX/2FZDuMo43NIWeeQk+a/WRMZ5andu0ie2XqlvtYf53jPGS\nLa4LAACYQ/tq0vDI6s+aGhgs+fbqzdVjtqqoTfSOpsv9nlrdeTz30uphXXd/0pLDqv+x7HUA\nAMAOs55L7JYcXv1w112ONs+uqs6o/rL639ULqrdXDxjj66p/qL6t+pHq1tWfVh/d/lIBAIBZ\n25+AdEPzlqbw89Lq58ax5OHjqGmT2JdUP7OdxQEAAPNjJwSkqr+ublf9++q06o5NzRourz5X\nfaB6ZVaOAABgR9spAammMPSScQAAAFzPvpo0AAAA7BhrDUjnN7W+vmT8XXXmsudWHmdubpkA\nAABbb62X2B21ynOHjmM1h+1fOQAAALOzr4D0zOo39+N9r96P1wAAAMzUvgLS1Qk7AADADqFJ\nAwAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAM\nAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAA\nDAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEA\nAAwCEgAAwCAgAQAADAISAADAICABAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAISAADAICAB\nAAAMAhIAAMAgIAEAAAwCEgAAwCAgAQAADAfPuoBtdK/qEdVJ1bHVEdVl1eer86pXV/97ZtUB\nAAAztxMC0k2qP64evIfzd6pOrX6+en31Y9XXt6c0AABgnuyEgHRWdVr1weol1blNq0ZXVIc1\nrSadUv1UU4h6dXX/mVQKAADM1KIHpPs0haM/qn6yunaVOf9cvbf6X9VLq8c0XY733u0pEQAA\nmBeL3qThnmP8lVYPR8tdU/3yitcBAAA7yKIHpEPGuNZ7ipbmHboFtQAAAHNu0QPSJ8Z4+hrn\nP2aMH9+CWgAAgDm36Pcgvbn6bPW71XHVC6tPrTLvDtUZ1X9puifpr7erQAAAYH4sekC6omnv\nozdWTx3HJdWF1ZVNl9IdV91szP/SmH/FtlcKAADM3KIHpKr3N+119F+bws8duy4QVe1quhTv\nrOr3qq9sd4EAAMB82AkBqabQ84vjOLJp76PDq8ub9kS6dHalAQAA82KnBKTlLm31+5AAAIAd\nbtG72AEAAKzZTlpBulfTPUgnNV1id0R1WdMldudVr67+98yqAwAAZm4nBKSbVH9cPXgP5+9U\nnVr9fPX66sda+8ayAADAAtkJAems6rTqg9VLqnObVo2uqA5rWk06pfqpphD16ur+M6kUAACY\nqUUPSPdpCkd/VP1kde0qc/65em/1v6qXVo9puhzvvdtTIgAAMC8WPSDdc4y/0urhaLlrql9u\nCkj3bGMB6cbV/9O0Ee1aHL+BzwIAADbJogekQ8a41nuKluatNdjsyZHVycs+f19uOsYDNvi5\nAADABix6QPrEGE+vnrOG+Y8Z48c3+LlfrB60jvn3qd5d7drg5wIAABuw6AHpzdVnq9+tjqte\n2OqbxN6hOqP6L033JP31dhUIAADMj0UPSFc07X30xuqp47ikurC6sulSuuOqm435Xxrzr9j2\nSgEAgJlb9IBU9f6mvY7+a1P4uWPXBaKaLmv7RFM78N+rvrLdBQIAAPNhJwSkmkLPL47jyKa9\njw6vLm/aE+nS2ZUGAADMi50SkJa7tNXvQwIAAHa4A2ddwBy4SXVCddCsCwEAAGZLQJruTfqX\npsvuAACAHWzRL7E7sH2HwKXzB3Xd73HtOAAAgB1k0VeQnl5dtY/jl8bcC5Y99/TtLhQAAJi9\nRV9BWvKN9rwidFjTfkhfb2r5XfZBAgCAHWnRV5A+1LQh7Merezftf7Ty+M0x986rPAcAAOwg\nix6Qzq7uUR1SnVf9fIv/nQEAgP20E8LCh6tTqmdXv1W9q7rDTCsCAADm0k4ISFXfqs6sfqCp\nnfeHqidWB8yyKAAAYL7slIC05G+r76r+rHpO9bbq1jOtCAAAmBs7LSDV1NHucdVDmhoz/MRM\nqwEAAObGTmnzvZrXV39X/UxTE4evz7YcAABg1nZyQKr6UvWMWRcBAADMh514iR0AAMCqBCQA\nAIBBQAIAABgEJAAAgEFAAgAAGAQkAACAQUACAAAYBCQAAIBBQAIAABgEJAAAgEFAAgAAGAQk\nAACAQUACAAAYBCQAAIBBQAIAABgEJAAAgEFAAgAAGAQkAACAQUACAAAYBCQAAIBBQAIAABgE\nJAAAgEFAAgAAGAQkAACAQUACAAAYBCQAAIBBQAIAABgEJAAAgEFAAgAAGAQkAACAQUACAAAY\nBCQAAIBBQAIAABgEJAAAgEFAAgAAGAQkAACAQUACAAAYBCQAAIBBQAIAABgEJAAAgEFAAgAA\nGAQkAACAQUACAAAYBCQAAIBBQAIAABgEJAAAgEFAAgAAGAQkAACA4eBZF7DNblrdrTq2OqK6\nrPp8dX71tRnWBQAAzIGdEpBOqZ5RnVYdtMr5q6u3Vk+vzt3GugAAgDmyEwLSo6uXN33XL1cf\naFo1uqI6rGk16eTqAU0B6vTqlTOpFAAAmKlFD0hHVy+oLqnOqM6udq0y78DqYdXzqxc2rSZd\nvE01AgAAc2LRmzQ8qDqq+unqta0ejqqurc6qHjfmP3BbqgMAAObKogekE8b4jjXOf/sYb7X5\npQAAAPNu0QPSUme6E9c4/7ZjvGQLagEAAObcogektzVdVve86ph9zL3FmLerOmeL6wIAAObQ\nojdp+ERT6HlC9enqdU1tvC+srqwOrY6v7lE9pDqyem71yVkUCwAAzNaiB6SqJ1UXVWc2tfA+\nfQ/zLm/aB+mZ21QXAAAwZ3ZCQLq2aZPYZzd1pzupOq46vCkUXVidV72x+vqMagQAAObATghI\nS75WvWIcAAAA17PoTRoAAADWbCetIFXdtLpbdWx1RHVZ9fnq/K5rCQ4AAOxQOyUgndJ0H9Jp\n1UGrnL+6emtTk4Zzt7EuAABgjuyEgPTo6uVN3/XL1QeaVo2uqA5rWk06uXpAU4A6vXrlTCoF\nAABmatED0tHVC6pLqjOqs5s2gl3pwOph1fOrFzatJl28TTUCAABzYtED0oOqo6rHVK/fy7xr\nq7Oqa6rXNrUD/5MNfO4B1X2aWomvxZ038FkAAMAmWfSAdMIY37HG+W8f4602+LknVn9T3Wid\nrztgg58LAABswKK3+V7qTHfiGuffdoyXbPBzP10d0hR41nJ873jdapf/AQAA22TRA9LbmkLH\n86pj9jH3FmPeruqcLa4LAACYQ4t+id0nmkLPE5pWdV7X1Mb7wurK6tDq+Ooe1UOqI6vnVp+c\nRbEAAMBsLXpAqnpSdVF1ZlML79P3MO/ypn2QnrlNdQEAAHNmJwSka5s2iX12U3e6k6rjmjrM\nXd60mnRe9cbq6zOqEQAAmAM7ISAt+Vr1inEAAABcz6I3aQCA/7+9O4+SrKrvAP7tnp6NWXQG\nRwQHlFFkUxbFfQERNCp6jscB44J4XBKNwRhxCybmaEhQohwOhqCCGiUGk5iAEXA3ETFIosYI\niOggIAqKGzAwDDMDnT/uq3RNUVX0vFfT1VP9+ZxTp6bffffWr6de33q/d++7BQDTJkECAACo\nzJUEaU3KfUgfSfLmJDv32G+XJJ9P8rIZigsAAJhF5sI9SE9PcmHKogwtJyY5LslnO/ZdnORZ\nSb45M6EBAACzyaiPIM1P8vGUxOfTSV6T5NQki5Kcn5IkAQAAJBn9EaSnJ9k9yTlJXt62/cNJ\nLkjy0SSbk/zDzIcGAADMNqM+gvTI6vmjHduvTvLkJD9I8okka2cyKAAAYHYa9QRpQfV8c5ey\nm5MckeTalBGko2YqKAAAYHYa9QTphup5dY/ym5I8I8nPU+5ROnImggIAAGanUU+Qvls9H95n\nn5+kjCTdkuTM7R4RAAAwa416gnRlkh8leWnKina9/DBl9OjWmQgKAACYnUZ9FbskeX2S/ZOs\nSnJjn/0uT/LMJEcnuWQG4gIAAGaZuZAgfal6TMe3qwcAADAHjfoUOwAAgGmTIAEAAFQkSAAA\nABUJEgAAQEWCBAAAUJEgAQAAVCRIAAAAFQkSAABARYIEAABQkSABAABUJEgAAAAVCRIAAEBF\nggQAAFCRIAEAAFQkSAAAABUJEgAAQEWCBAAAUJEgAQAAVCRIAAAAFQkSAABARYIEAABQmRh2\nAAAAzApHJ3lbwzbWDCIQGCYJEgAASXLgmjVrHnPsscfWbuDUU08dYDgwHBIkAACSJLvsskvW\nrl1bu/6ZZ545wGhgONyDBAAAUJEgAQAAVCRIAAAAFQkSAABAxSINAAA7vl2SHJdkrEEbTxxQ\nLLBDkyABAOz4jly4cOF7Dz744NoNXHnllQMMB3ZcEiQAgB3f2IoVK3LWWWfVbuDlL3/5AMOB\nHZd7kAAAACoSJAAAgIoECQAAoCJBAgAAqEiQAAAAKhIkAACAigQJAACgIkECAACoSJAAAAAq\nEiQAAICKBAkAAKAiQQIAAKhIkAAAACoSJAAAgIoECQAAoCJBAgAAqEiQAAAAKhIkAACAigQJ\nAACgMjHsAAAAyJIkCxrWBwZAggQAMFx7JflBzOwZNWNJDkiz8+0HDygWtoEECQBguO6XZPzj\nH/94Fi1aVKuBs88+O5dffvlgo6KpZyb5/LCDYNtJkAAAmlmYZKcG9ZcnyT777JOddqrXzIoV\nKxq8PNvJoqVLl+YLX/hC7QZOOOGEbNiwYYAhMR0SJACA+hYk+WWSZcMOhNlp+fLltetOTDhV\nHwb/6wAA9S1Isuzkk0/OmjVrajVw8cUX54wzzhhsVEBtEiQAgIb23HPP7LfffrXqXnPNNQOO\nBmhCggQAwEiZnJxMkj9O8qIGzdya5MQkWwYREzsOCRIAACNlcnIyhxxyyFErV66sVX/Dhg25\n5JJLkuS0JDcOMjZmPwkSAAAj57WvfW0e//jH16p7/fXX56ijjhpwROwoJEgAwFy2onrU1WR5\nb2apzZs3t/55YpLbazaz12CiYabNxQRpIsnKlA5tQ5LfxNxSANgR7ZHkEQ3b+EjVDvy/X/zi\nF0mSww477PULFiyo1ca6dety8803DzIsZshcSZCWJXl9khcmOSBlSc6Wu5JcnuTTSc5I/asE\nAMD0vTLJOxq28aAMYATnLW95Sw4//PBadW+55Za8+MUvbhoCs9Q73/nOrFq1qlbd008/Peee\ne+6AI2ImzIUE6VFJPpfkwdXPG5NcVz0vSulcD6kexyd5dkrCBABsPwftu+++a4455pjaDZxy\nyik59thjc/zxx9cP4qCDsmLFiqxevbpW/cWLF9d+bWB2GvUEaWGS85LsmuT0JGcnuSLJZNs+\nYylJ1GuS/EGSf02yf5JNMxopAL0sSfL4JOMN2nhakgNTZg3UtVOSLyW5skEbW5J8PcndDdp4\nbJL7NaifJDcn+V6D+gcn+ZOGMRy8evXqrF27tnYDp512WsMQAO5t1BOkZyZ5WJI3J3l/j30m\nUz4kjk/ykySnVPUumIkAZ9jeSQ5t2MZkks8m+XnzcEiyJskRDdvYI8n6JL9t0MbdKdNMb20Y\nSxMTSV6SMrLbxLokX21Qf/eUkeSmvpzkxwNoZ5hWpPy9N3lPxlOOr7MatHFEkqMb1E+S7LXX\nXnnqU59au/4555yTzZs3P7dpHCnHaN2/tYmURK+pySTfaVB/1YoVK/Z4wQteULuB888/v8HL\nA2w/Y8MOYDt7a5L3JlmV5FfT2H9VylW1tyb56wavu2eSyzL9BHQi5T6pBUk238e+TXw4ZaSs\nqTtTpijWtSxlgYy6V1DHqjbWZ+vRwG0xL+Wq9G016yfJ/JRRyib3rS1O84RgUO5Is5HT5Sn/\nF/fUrD+vaqOpe9Is0VuU8r40tTHlb6WupSmjHU36hOUp72vdv7XxNB+pGIh58+ZlyZIltevf\nfvvtmTdvXqPpUOvXr8+CBQuycOHC2m3cdluTLmfK0qVLMz5eb0Bt06ZN2bixSRdejI+PZ+nS\npbXr33HHHRkfH58V78nixYszf/78WvUnJyezfv36Ru/Jli1bsmHDhixbtixjY/VOzTZu3Jgt\nW7Y0fk/Gxsay0071b+u6/fbbMzExkUWL6n+03XbbbVm0aFHqLo7QamPJkiWZN29erfqt96TJ\n+3rXXXdl06ZNWbZsWa36Sfk+psnJyUb935133pnNmzd/JMmrazcyx4x6gvSmlJGjhya5fhr7\nPzTJtSnfvNxk3H48ZTrHdBOksSQPTPLJBq85HbumTB9sYreUk887GrSxJmW0rsnqgXsl+VGD\n+uMpiew1DdpYlJJU39CgjfunJFq/bNDGg1Lej/UN2nhIkpvSLEF6eMr/Z92kdSxlxHddgxgW\npBzn0/l772VZSvLcZJR0VUpic0uDNnZPOS6anMmuSbnnsm7SmjT/W5tI+V2ubdDGkpRErcmX\nNba+LfI3DdrQ/03R/21N/zdF/zdltvR/SZkafFPDNhgRT0nprM7KfSeDYylLfU4mefJ2jgsA\nAGDGjSX5WkrSc2nK0OKBKVc3llfPB6VMO7us2u9rGf2RNQAAYI7aOclXUpKf+3p8pdofAACY\ng+bSSMmRKV8U++iUeeSLU26gvjFlJZ9/SVm+FQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAICZNDbsAIDGrkmyZthBAADbxU1Jdht2\nEHPJxLADABq7Icl/JPnbIccB3Rya5F1JDhtyHNDLhUnOSHLRsAOBLl6R5InDDmKukSDBjm9T\nytWlbw87EOhidZK74/hk9tqU5Lo4RpmdjkyyedhBzDXjww4AAABgtpAgAQAAVCRIAAAAFQkS\nAABARYIEAABQkSABAABUJEgAAAAVCRIAAEBFggQAAFCRIMGOb1N8yzaz16bqAbOVY5TZzPEJ\nUMMDkywddhDQw3iShw47COhjjyQTww4CelicZNdhBwEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACwnYwNOwDgXtYk2aP699eSTPbZ9wFJHpLkriQ/\nSXLbNNpfmeThVZ0fVM/QzYIkT+pTfnOS7/cpX5RknyTzk6xL8ttpvGadOsxdj0iyW5/y/0qy\noUfZ/ZLsmXKc/zjJr6b5mrsnWZ3kN0l+mP59NHNb0z60ZWWSA6p//2/694v6UGDkrExyY8oH\n7mSSiR77PSfJd9r2m0xyT5ILMtWJdto5yaeSbGmrc2uSd8TFErp7ULY+xjofn+pRbzzJu5Os\nb9t3c5JzUk5KB1UHPpj+x+g+HfvfL8lfJLmiy75fT3JIn9c6MMllHXWuS/K8gfwmjKK6fWin\nz7fVOaLHPvrQAep18gUMxwdTkpU7kyzusc+zUhKhDUnek+RbKVeMnp/kmCRPTbJ/kp+21Zmf\n5HNJHpvk/CSfTLI8yQlJTqrq/9lgfxVGyBVJ3t9l+7U99j81yR+lXOk8LckdSY5L8rKU0dGn\npyT0TetAy0lJrumy/aaOn/dM8qcpI0ZnJLk6yZIkz07ytCQXp/Sh3+6otybJV1NONN+X5Bsp\nI/EnJjkvyXOTfGEAvwejaVv70HavT3JkkutTZoz0og8FRtLLUq74PC/JLek9gvSZquy4LmX/\nVJW9sWP766rt53dsX5lyArEp5cQB2rWufn5+G+rsn+TulCmfy9u2j2XqKuhLB1AHkqkRpCdM\nc//dkjyzR9kZVVuf6VL2j+netz6p2v7DGInn3ur0oe0ekZLovC/l87vXCJI+FBhJu6fME/5Y\n9XO/BOnf0/uE4F1V2Rs6tv/nNOqcuM1RM+rqfLj/VVXn7V3Knl6VfXEAdSDZ9gSpn/2rttZ1\nbF+WZGPKlOSFXeq17hXtd68Jc1OTBGkiZUrnVSmzPPolSPpQYOSMJflykhsyNU+4X4L0zkyN\nBrXPK94n5erRLdn6puX5KQsxbEgyr0t7h1ft/Vvt34BR1fpw/1mSS1OmJX0ryd8k2bdHnVYC\n3+1kcUHKPXC3DqAOJFMJ0sUp0+WuTnJRklen9H3b4glVW9/o2H5o+p9gvrsqf9M2vh6jr04f\n2vLnKX3f46qf+yVI+lBg5LwxpWNrn/bRL0FanOTTVfkdSb6XMr1jc8rUkM5O96GZmgLSzcOq\n8strRc8oa32435Iyr/1/MnUD8F1Jju1S57qqvNfKYjdU5Ts3rANJSZDuTjnx/GbKfRr3pBwv\nl2bbbk4/r6r3mo7tr6i2f7hHvVdV5R/YhtdibqjThyZlsZDNSf6ybVu/BOm66EOBEbJvyoIM\nH+zY3i9BSspo0UUpCdJ3Upbr3pKS5HTOr39k1dZ3erS1S1U+nZtFmVt2ShlhbB95XJhyk3tr\nhaS9O+r8qipbnu6uqsrbbzauUweScnV9145tB6dcOJpM8olptvP2av8vpawG1u4Pq7JTe9R9\nUVX+sR7lzF11+tDFKX3e91JGf1r6JUj6UGBkzE9ZKenaJEs7yvolSL+TMl3uvCQr2rbvn7KK\n05YkT27bvnfV1vd6xLG6Kr9628JnjvtoynHzno7tN1XbV/aot64qbz+prVMH+lmT0hduSu+T\nxpZWAvTtbN2ntvx+VX56j/qtBXY+VCtS5qpefegHUo7bgzu290uQ9KED1nmVBJg5r0vy6CR/\nlzKcfljbo5UYHVr93H5j8IdTriq9Ilt/AdyVSd6ccqXqLW3bW19+uKpHHA+onn+9beEzx11U\nPe/XsX06x9s92frYrVMH+vlxysj6/JTluHt5a8oJ6X+nnHh2O8b0oWwPHvbh8gAAA/VJREFU\n3frQA1OW9b4wZXroYW2PB7Ttc1jK7I8WfSgwMk5K/y+Qa3+srursWv18Q482W6NFV3Rsb11d\n6tZ5Hpv+8+uhm9ZV8/M6tp9bbV/bpc7u6X4/XJ06cF9aV80P7FH+F1X5Jek/ytSrX235SFX+\nknphMkd160OPyPTPC17WVk8fOmC+KBaG559TrnB286GUucvHpVz1+U21fUOmhslXtm1vOaB6\n/kXH9ouSvDLJC3Pv+53Wtu0D03V09fzdju0XJfndlOPq0x1lvY61OnWgn8ekTLO7M/c+MRxL\n+SLNN6R8+evzU+7n7OXqlOnL+6fcN3pVW9nCJEelTOezjDLbolsfekV6L9zwxpTj+j0pM0ba\nV1vUhwJzwnS+B+nCbD0idHDKCk6TSV7bUWe/lJtBf5nkoLbtr672v7LHazG3vS3JMUnu37Zt\njyRnpxw365M8uKPOwpTpTfdk6w/6x6ZM79iYe38pcZ06sHeSk1MSl1b/tSDJC1K+8qDbwgrj\nmRrxuSjl+2Wm4/cytTJea6rTgkzdR+L+I7qp04f20u8eJH0oMCf0S5AenrKwQ2uZ0O9XP7eW\ntv37dP++o1elJEl3p6x210qmfp5730cCSXJBpqZz3J7ktraff5myOlM3j87UqkrXplwVvSfl\nxuMXD7AOc9shmToe7045Ju9u23Zutl4FLEmObCv/WcryyJ2PbgvWjGUqsdqQslzzrzOVNHUu\ntANJ/T60m34JUqIPHaixYQcAdHVBygfuM1I+8DstSZnv/rSUGzW3pEwBOT/JV/q0+6iURGnv\nlE7zspQrn24uppu9Uj6MH53yfR4LU6ZvfiPl5LPfFw8+MOWq++NSEv2rUq6aXjXgOsxdC1OO\nz6ekXB3fOWVK3Q9S7uu4tEudxyU55T7a3ZR7f11Cy3NSpiw9OOXK/BdTlhLfso2xMzc06UM7\nnZRyrJ+QsuJiN/pQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAALa/\nM5Ocn+Rhww4EAABg2NYlmUxyyLADAWB2mxh2AAAwAy5L8tMk64cdCAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEPxf9UO+9XBRfr3AAAAAElFTkSu\nQmCC",
      "text/plain": [
       "Plot with title “Histogram of .”"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "## per protein cases count \n",
    "# make table with ncases / ncontrols per protein - will use later\n",
    "del_per_prot <- data_qc[,-c(1:6,8:30)] %>% group_by(delirium_incident) %>% summarise(across(where(is.numeric), ~ sum(!is.na(.))))\n",
    "del_per_prot\n",
    "# hist\n",
    "del_per_prot[2,-1] %>% t %>% hist(.,breaks=50)\n",
    "# summary\n",
    "print(\"No. of delirium cases summary:\")\n",
    "del_per_prot[,-1] %>% t %>% .[,2] %>% summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be07383a",
   "metadata": {},
   "source": [
    "### Exclude proteins with <150 delirium cases:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "7771271e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteins to be excluded:\n"
     ]
    }
   ],
   "source": [
    "# transpose del_per_prot table & filter proteins with < 150 cases\n",
    "prot_names_low_cases <-\n",
    "    tibble(prot = colnames(del_per_prot[,-1]), \n",
    "           n_cases = as.numeric(del_per_prot[2,-1]), \n",
    "           n_controls = as.numeric(del_per_prot[1,-1])) %>%\n",
    "    filter(n_cases < 150) %>% pull(prot)\n",
    "\n",
    "cat(paste0(\"Proteins to be excluded:\", \"\\n\", paste(prot_names_low_cases, collapse = \"\\n\")))\n",
    "\n",
    "# filter out proteins to be excluded\n",
    "data_qc <- data_qc %>% select(!any_of(prot_names_low_cases))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6641d0",
   "metadata": {},
   "source": [
    "### **Inverse-rank normalisation of proteins:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "35e49b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein names\n",
    "prot_names <- names(data_qc)[32:ncol(data_qc)]  # skip all covariates (check first)\n",
    "\n",
    "# i.r.n.t. for every protein in prot_names\n",
    "# data_int <-\n",
    "data_qc <- \n",
    "data_qc %>% mutate(across(all_of(prot_names), ~ (qnorm((rank(.,na.last=\"keep\")-0.5)/sum(!is.na(.))))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5a024d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#test irnt\n",
    "#a <- data_int %>% pull(ABHD14B) \n",
    "#b <- data %>% mutate(ABHD14B = qnorm((rank(ABHD14B,na.last=\"keep\")-0.5)/sum(!is.na(ABHD14B))) ) %>% pull(ABHD14B)\n",
    "#data.frame(a,b) %>% filter(is.na(a))\n",
    "#all(a==b,na.rm = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec459fde",
   "metadata": {},
   "source": [
    "### **Check significant covariates:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "911f260e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "delirium_incident ~ sex + age_at_collection + BMI + GeneticArray + \n",
      "    UKB_centre + protein_batch + Sample_age + e4_count\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“glm.fit: fitted probabilities numerically 0 or 1 occurred”\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "glm(formula = formula0, family = \"binomial\", data = data_qc)\n",
       "\n",
       "Coefficients:\n",
       "                    Estimate Std. Error z value Pr(>|z|)    \n",
       "(Intercept)       -5.222e+02  2.039e+02  -2.561 0.010432 *  \n",
       "sex                3.674e-01  8.858e-02   4.148 3.35e-05 ***\n",
       "age_at_collection  1.732e-01  9.312e-03  18.603  < 2e-16 ***\n",
       "BMI                4.563e-02  8.986e-03   5.078 3.82e-07 ***\n",
       "GeneticArray       3.338e-02  1.409e-01   0.237 0.812779    \n",
       "UKB_centre         4.543e-02  1.839e-02   2.471 0.013490 *  \n",
       "protein_batch     -5.684e-02  3.120e-02  -1.822 0.068444 .  \n",
       "Sample_age         1.191e-03  3.333e-04   3.574 0.000351 ***\n",
       "e4_count           4.210e-01  7.776e-02   5.414 6.16e-08 ***\n",
       "---\n",
       "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
       "\n",
       "(Dispersion parameter for binomial family taken to be 1)\n",
       "\n",
       "    Null deviance: 5509.1  on 32641  degrees of freedom\n",
       "Residual deviance: 4880.9  on 32633  degrees of freedom\n",
       "  (10 observations deleted due to missingness)\n",
       "AIC: 4898.9\n",
       "\n",
       "Number of Fisher Scoring iterations: 8\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "PC1to20   <- paste0(\"PC_\", seq(1,20), collapse=\" + \")\n",
    "formula0  <- \n",
    "as.formula(\n",
    "    paste(\n",
    "    \"delirium_incident ~ sex + age_at_collection + BMI + GeneticArray + UKB_centre + protein_batch + Sample_age + e4_count\",\n",
    "    #PC1to20,\n",
    "    sep=\" + \"))\n",
    "print(formula0)\n",
    "mod0 <- glm(formula = formula0, data = data_qc, family=\"binomial\")\n",
    "summary(mod0)\n",
    "## keep sex, age_at_collection, BMI  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41f4ca70",
   "metadata": {},
   "source": [
    "### **Export clean data table (no extra covariates):**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e93b5377",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_clean <- data_qc %>% select(c(\"IID\",\"delirium_incident\",\"sex\",\"age_at_collection\",\"BMI\",\"age\",\"e4_count\"),names(data_qc)[32:ncol(data_qc)])\n",
    "file <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/data/data_int_clean.csv\"\n",
    "write.table(data_clean, file=file, sep = \",\", row.names = F, col.names = T, quote = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c7a2dcd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## LOAD CLEAD DATA\n",
    "file    <- \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/data/data_int_clean.csv\"\n",
    "data_qc <- fread(file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73356182",
   "metadata": {},
   "source": [
    "### **Logistic Regression on full set:**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc871776-b24c-4125-9e2a-9654e3d6ae43",
   "metadata": {},
   "source": [
    "#### ***delirium ~ age + sex + BMI + protein_i***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60ed34c2-48b5-4c0c-b4b6-462a13371bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# real\n",
    "# get protein names\n",
    "noprot_names <- c('IID','delirium_incident','sex','age_at_collection','BMI','age','e4_count')\n",
    "prot_names   <- names(data_qc)[names(data_qc)!=noprot_names]  # skip all covariates (check first)\n",
    "# create empty output sumstats table\n",
    "prot_sumstats <- data.frame(PROTEIN=numeric(), \n",
    "                            BETA=numeric(), \n",
    "                            STDERR=numeric(), \n",
    "                            PVALUE=numeric(), \n",
    "                            #PROTxE4=numeric(),\n",
    "                            #PROTxE4_STDERR=numeric(),\n",
    "                            #PROTxE4_PVALUE=numeric(),\n",
    "                            NCASES=numeric(), \n",
    "                            NCONTROLS=numeric()\n",
    "                           )\n",
    "# assign dataframe \n",
    "data_for_loop = data_qc\n",
    "\n",
    "# progress bar\n",
    "pb = txtProgressBar(min = 0, max = length(prot_names), initial = 0) \n",
    "\n",
    "# run model: \"incident delirium ~ covariates + protein\" for each protein in prot_names\n",
    "for (i in 1:length(prot_names)) {\n",
    "    #print(paste0(\"analyzing protein \", prot_names[i], \" ...\"))\n",
    "    ## create formula string\n",
    "    formula_i  <- as.formula(\n",
    "        paste(\n",
    "            \"delirium_incident ~ sex + age_at_collection + BMI + \", prot_names[i]))\n",
    "    ## logistic reg model: delirium ~ coviates + protein_i\n",
    "    mod_i   <- glm(formula = formula_i, data = data_for_loop, family=\"binomial\",) %>% suppressWarnings()\n",
    "    ## extract coefficients: beta; std error; p-value\n",
    "    coefs_i <- summary(mod_i)$coefficients[prot_names[i],c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")] \n",
    "    ## append coefs in output sumstats table\n",
    "    prot_sumstats[i,c(\"BETA\",\"STDERR\",\"PVALUE\")] <- coefs_i\n",
    "    ## append protein_i name in output sumstats table\n",
    "    prot_sumstats[i,\"PROTEIN\"] <- prot_names[i]\n",
    "    # ## append e4 interaction sumstats\n",
    "    # prot_x_e4_i       <- paste0(\"e4_count:\", prot_names[i])\n",
    "    # prot_x_e4_coefs_i <- summary(mod_i)$coefficients[prot_x_e4_i,c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")]\n",
    "    # prot_sumstats[i,c(\"PROTxE4\",\"PROTxE4_STDERR\",\"PROTxE4_PVALUE\")] <- prot_x_e4_coefs_i\n",
    "    ## append protein_i no. of cases + controls (using del_per_prot_train table)\n",
    "    ncases    <- del_per_prot[2,prot_names[i]] # 2nd row: cases\n",
    "    ncontrols <- del_per_prot[1,prot_names[i]] # 1st row: cases\n",
    "    prot_sumstats[i,c(\"NCASES\",\"NCONTROLS\")] <- c(ncases, ncontrols)\n",
    "    #print(\"Done\")\n",
    "    setTxtProgressBar(pb,i)\n",
    "    cat(paste0(\"\\rAnalysed \", i, \" out of \",  length(prot_names), \" proteins (\", round((i / length(prot_names)*100),2), \"%) ...\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a6e2f8-a6fc-429f-af51-da80dd7dcbe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_sumstats_annot <- inner_join(x=prot_sumstats, y=assay, by=c(\"PROTEIN\"=\"Assay\")) \n",
    "# make UniProt var to merge \n",
    "prot_sumstats_annot <- prot_sumstats_annot %>% mutate(UniProt2 = gsub(\"-.*\",\"\", UniProt)) %>% mutate(UniProt2 = gsub(\"_.*\",\"\", UniProt2))\n",
    "\n",
    "prot_sumstats_annot_pos <- \n",
    "left_join(x=prot_sumstats_annot, y=pos[,-c(\"Assay_Target\",\"Gene_symbol\")], by=c(\"UniProt2\"=\"UniProt\")) %>% \n",
    "    mutate(Gene_CHROM = gsub(\"X\",\"23\",Gene_CHROM)) %>%\n",
    "    mutate(Gene_CHROM = as.numeric(Gene_CHROM)) %>%\n",
    "    mutate(Gene_start = as.numeric(Gene_start)) %>%\n",
    "    mutate(Gene_end = as.numeric(Gene_end)) %>%\n",
    "    arrange(Gene_CHROM, Gene_start)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab686b3e-3f1f-495f-bcf4-1a9e90be96a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output \n",
    "dim(prot_sumstats_annot_pos)\n",
    "prot_sumstats_annot_pos %>% arrange(PVALUE) %>% filter(PVALUE < 0.05/nrow(prot_sumstats_annot_pos)) %>% nrow\n",
    "prot_sumstats_annot_pos %>% arrange(PVALUE) %>% filter(PVALUE < 0.05/nrow(prot_sumstats_annot_pos)) %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52385588-7c4b-461b-9311-cb8bd6d19085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write output \n",
    "write.table(prot_sumstats_annot_pos, \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/output_pwas_models12/del_model1_glm_int_fullset_sumstats.txt\", sep = \",\", row.names = F, col.names = T, quote = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bafe74d",
   "metadata": {},
   "source": [
    "#### ***delirium ~ age + sex + BMI + protein_i + APOE-e4 x protein interaction***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "74ac38b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========================================================================================================================================================================================================"
     ]
    }
   ],
   "source": [
    "# real\n",
    "# get protein names\n",
    "prot_names <- names(data_qc)[32:ncol(data_qc)]  # skip all covariates (check first)\n",
    "# create empty output sumstats table\n",
    "prot_sumstats <- data.frame(PROTEIN=numeric(), \n",
    "                            BETA=numeric(), \n",
    "                            STDERR=numeric(), \n",
    "                            PVALUE=numeric(), \n",
    "                            PROTxE4=numeric(),\n",
    "                            PROTxE4_STDERR=numeric(),\n",
    "                            PROTxE4_PVALUE=numeric(),\n",
    "                            NCASES=numeric(), \n",
    "                            NCONTROLS=numeric()\n",
    "                           )\n",
    "# progress bar\n",
    "pb = txtProgressBar(min = 0, max = length(prot_names), initial = 0) \n",
    "\n",
    "# run model: \"incident delirium ~ covariates + protein\" for each protein in prot_names\n",
    "for (i in 1:length(prot_names)) {\n",
    "    #print(paste0(\"analyzing protein \", prot_names[i], \" ...\"))\n",
    "    ## create formula string\n",
    "    formula_i  <- as.formula(\n",
    "        paste(\n",
    "            \"delirium_incident ~ sex + age_at_collection + BMI + e4_count*\", prot_names[i]))\n",
    "    ## logistic reg model: delirium ~ coviates + protein_i\n",
    "    mod_i   <- glm(formula = formula_i, data = data_qc, family=\"binomial\",) %>% suppressWarnings()\n",
    "    ## extract coefficients: beta; std error; p-value\n",
    "    coefs_i <- summary(mod_i)$coefficients[prot_names[i],c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")] \n",
    "    ## append coefs in output sumstats table\n",
    "    prot_sumstats[i,c(\"BETA\",\"STDERR\",\"PVALUE\")] <- coefs_i\n",
    "    ## append protein_i name in output sumstats table\n",
    "    prot_sumstats[i,\"PROTEIN\"] <- prot_names[i]\n",
    "    ## append e4 interaction sumstats\n",
    "    prot_x_e4_i       <- paste0(\"e4_count:\", prot_names[i])\n",
    "    prot_x_e4_coefs_i <- summary(mod_i)$coefficients[prot_x_e4_i,c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")]\n",
    "    prot_sumstats[i,c(\"PROTxE4\",\"PROTxE4_STDERR\",\"PROTxE4_PVALUE\")] <- prot_x_e4_coefs_i\n",
    "    ## append protein_i no. of cases + controls (using del_per_prot_train table)\n",
    "    ncases    <- del_per_prot[2,prot_names[i]] # 2nd row: cases\n",
    "    ncontrols <- del_per_prot[1,prot_names[i]] # 1st row: cases\n",
    "    prot_sumstats[i,c(\"NCASES\",\"NCONTROLS\")] <- c(ncases, ncontrols)\n",
    "    #print(\"Done\")\n",
    "    setTxtProgressBar(pb,i)\n",
    "    #cat(paste0(\"\\rAnalysed \", i, \" out of \",  length(prot_names), \" proteins (\", round((i / length(prot_names)*100),2), \"%) ...\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "50098b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_sumstats_annot <- inner_join(x=prot_sumstats, y=assay, by=c(\"PROTEIN\"=\"Assay\")) \n",
    "# make UniProt var to merge \n",
    "prot_sumstats_annot <- prot_sumstats_annot %>% mutate(UniProt2 = gsub(\"-.*\",\"\", UniProt)) %>% mutate(UniProt2 = gsub(\"_.*\",\"\", UniProt2))\n",
    "\n",
    "prot_sumstats_annot_pos <- \n",
    "left_join(x=prot_sumstats_annot, y=pos[,-c(\"Assay_Target\",\"Gene_symbol\")], by=c(\"UniProt2\"=\"UniProt\")) %>% \n",
    "    mutate(Gene_CHROM = gsub(\"X\",\"23\",Gene_CHROM)) %>%\n",
    "    mutate(Gene_CHROM = as.numeric(Gene_CHROM)) %>%\n",
    "    mutate(Gene_start = as.numeric(Gene_start)) %>%\n",
    "    mutate(Gene_end = as.numeric(Gene_end)) %>%\n",
    "    arrange(Gene_CHROM, Gene_start)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "16caafd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>2919</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 2919\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 2919\n",
       "2. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 2919   15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "39"
      ],
      "text/latex": [
       "39"
      ],
      "text/markdown": [
       "39"
      ],
      "text/plain": [
       "[1] 39"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROTEIN</th><th scope=col>BETA</th><th scope=col>STDERR</th><th scope=col>PVALUE</th><th scope=col>PROTxE4</th><th scope=col>PROTxE4_STDERR</th><th scope=col>PROTxE4_PVALUE</th><th scope=col>NCASES</th><th scope=col>NCONTROLS</th><th scope=col>UniProt</th><th scope=col>Panel</th><th scope=col>UniProt2</th><th scope=col>Gene_CHROM</th><th scope=col>Gene_start</th><th scope=col>Gene_end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>GDF15</td><td>0.4423138</td><td>0.06574961</td><td>1.729034e-11</td><td> 0.0170975958</td><td>0.08780006</td><td>0.8456017</td><td>532</td><td>31884</td><td>Q99988</td><td>Cardiometabolic</td><td>Q99988</td><td>19</td><td>18374731</td><td>18389176</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>AREG </td><td>0.3627137</td><td>0.05944031</td><td>1.046507e-09</td><td>-0.0002928142</td><td>0.08522372</td><td>0.9972586</td><td>533</td><td>31945</td><td>P15514</td><td>Oncology       </td><td>P15514</td><td> 4</td><td>74445136</td><td>74455005</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>WFDC2</td><td>0.3716657</td><td>0.06143184</td><td>1.448009e-09</td><td> 0.0987329458</td><td>0.08681397</td><td>0.2554157</td><td>513</td><td>30513</td><td>Q14508</td><td>Oncology       </td><td>Q14508</td><td>20</td><td>45469753</td><td>45481532</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>IL6  </td><td>0.3412059</td><td>0.05893360</td><td>7.052633e-09</td><td>-0.0064074146</td><td>0.08030888</td><td>0.9364085</td><td>533</td><td>31945</td><td>P05231</td><td>Oncology       </td><td>P05231</td><td> 7</td><td>22725884</td><td>22732002</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>TFF3 </td><td>0.3444147</td><td>0.06101244</td><td>1.651901e-08</td><td>-0.0550779872</td><td>0.08438667</td><td>0.5139588</td><td>527</td><td>31636</td><td>Q07654</td><td>Cardiometabolic</td><td>Q07654</td><td>21</td><td>42311667</td><td>42315409</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PLAUR</td><td>0.3223726</td><td>0.05834944</td><td>3.297437e-08</td><td> 0.0303127498</td><td>0.08344665</td><td>0.7164114</td><td>520</td><td>31512</td><td>Q03405</td><td>Inflammation   </td><td>Q03405</td><td>19</td><td>43646095</td><td>43670547</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & PROTEIN & BETA & STDERR & PVALUE & PROTxE4 & PROTxE4\\_STDERR & PROTxE4\\_PVALUE & NCASES & NCONTROLS & UniProt & Panel & UniProt2 & Gene\\_CHROM & Gene\\_start & Gene\\_end\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & GDF15 & 0.4423138 & 0.06574961 & 1.729034e-11 &  0.0170975958 & 0.08780006 & 0.8456017 & 532 & 31884 & Q99988 & Cardiometabolic & Q99988 & 19 & 18374731 & 18389176\\\\\n",
       "\t2 & AREG  & 0.3627137 & 0.05944031 & 1.046507e-09 & -0.0002928142 & 0.08522372 & 0.9972586 & 533 & 31945 & P15514 & Oncology        & P15514 &  4 & 74445136 & 74455005\\\\\n",
       "\t3 & WFDC2 & 0.3716657 & 0.06143184 & 1.448009e-09 &  0.0987329458 & 0.08681397 & 0.2554157 & 513 & 30513 & Q14508 & Oncology        & Q14508 & 20 & 45469753 & 45481532\\\\\n",
       "\t4 & IL6   & 0.3412059 & 0.05893360 & 7.052633e-09 & -0.0064074146 & 0.08030888 & 0.9364085 & 533 & 31945 & P05231 & Oncology        & P05231 &  7 & 22725884 & 22732002\\\\\n",
       "\t5 & TFF3  & 0.3444147 & 0.06101244 & 1.651901e-08 & -0.0550779872 & 0.08438667 & 0.5139588 & 527 & 31636 & Q07654 & Cardiometabolic & Q07654 & 21 & 42311667 & 42315409\\\\\n",
       "\t6 & PLAUR & 0.3223726 & 0.05834944 & 3.297437e-08 &  0.0303127498 & 0.08344665 & 0.7164114 & 520 & 31512 & Q03405 & Inflammation    & Q03405 & 19 & 43646095 & 43670547\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 15\n",
       "\n",
       "| <!--/--> | PROTEIN &lt;chr&gt; | BETA &lt;dbl&gt; | STDERR &lt;dbl&gt; | PVALUE &lt;dbl&gt; | PROTxE4 &lt;dbl&gt; | PROTxE4_STDERR &lt;dbl&gt; | PROTxE4_PVALUE &lt;dbl&gt; | NCASES &lt;dbl&gt; | NCONTROLS &lt;dbl&gt; | UniProt &lt;chr&gt; | Panel &lt;chr&gt; | UniProt2 &lt;chr&gt; | Gene_CHROM &lt;dbl&gt; | Gene_start &lt;dbl&gt; | Gene_end &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | GDF15 | 0.4423138 | 0.06574961 | 1.729034e-11 |  0.0170975958 | 0.08780006 | 0.8456017 | 532 | 31884 | Q99988 | Cardiometabolic | Q99988 | 19 | 18374731 | 18389176 |\n",
       "| 2 | AREG  | 0.3627137 | 0.05944031 | 1.046507e-09 | -0.0002928142 | 0.08522372 | 0.9972586 | 533 | 31945 | P15514 | Oncology        | P15514 |  4 | 74445136 | 74455005 |\n",
       "| 3 | WFDC2 | 0.3716657 | 0.06143184 | 1.448009e-09 |  0.0987329458 | 0.08681397 | 0.2554157 | 513 | 30513 | Q14508 | Oncology        | Q14508 | 20 | 45469753 | 45481532 |\n",
       "| 4 | IL6   | 0.3412059 | 0.05893360 | 7.052633e-09 | -0.0064074146 | 0.08030888 | 0.9364085 | 533 | 31945 | P05231 | Oncology        | P05231 |  7 | 22725884 | 22732002 |\n",
       "| 5 | TFF3  | 0.3444147 | 0.06101244 | 1.651901e-08 | -0.0550779872 | 0.08438667 | 0.5139588 | 527 | 31636 | Q07654 | Cardiometabolic | Q07654 | 21 | 42311667 | 42315409 |\n",
       "| 6 | PLAUR | 0.3223726 | 0.05834944 | 3.297437e-08 |  0.0303127498 | 0.08344665 | 0.7164114 | 520 | 31512 | Q03405 | Inflammation    | Q03405 | 19 | 43646095 | 43670547 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROTEIN BETA      STDERR     PVALUE       PROTxE4       PROTxE4_STDERR PROTxE4_PVALUE NCASES NCONTROLS UniProt Panel           UniProt2 Gene_CHROM Gene_start Gene_end\n",
       "1 GDF15   0.4423138 0.06574961 1.729034e-11  0.0170975958 0.08780006     0.8456017      532    31884     Q99988  Cardiometabolic Q99988   19         18374731   18389176\n",
       "2 AREG    0.3627137 0.05944031 1.046507e-09 -0.0002928142 0.08522372     0.9972586      533    31945     P15514  Oncology        P15514    4         74445136   74455005\n",
       "3 WFDC2   0.3716657 0.06143184 1.448009e-09  0.0987329458 0.08681397     0.2554157      513    30513     Q14508  Oncology        Q14508   20         45469753   45481532\n",
       "4 IL6     0.3412059 0.05893360 7.052633e-09 -0.0064074146 0.08030888     0.9364085      533    31945     P05231  Oncology        P05231    7         22725884   22732002\n",
       "5 TFF3    0.3444147 0.06101244 1.651901e-08 -0.0550779872 0.08438667     0.5139588      527    31636     Q07654  Cardiometabolic Q07654   21         42311667   42315409\n",
       "6 PLAUR   0.3223726 0.05834944 3.297437e-08  0.0303127498 0.08344665     0.7164114      520    31512     Q03405  Inflammation    Q03405   19         43646095   43670547"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# output \n",
    "dim(prot_sumstats_annot_pos)\n",
    "prot_sumstats_annot_pos %>% arrange(PVALUE) %>% filter(PVALUE < 0.05/nrow(prot_sumstats_annot_pos)) %>% nrow\n",
    "prot_sumstats_annot_pos %>% arrange(PVALUE) %>% filter(PVALUE < 0.05/nrow(prot_sumstats_annot_pos)) %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "56d02f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write output \n",
    "write.table(prot_sumstats_annot_pos, \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/output_pwas_models12/del_model1_glm_int_fullset_e4int_sumstats.txt\", sep = \",\", row.names = F, col.names = T, quote = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f908b7c2",
   "metadata": {},
   "source": [
    "#### ***delirium ~ age + sex + BMI + protein_i + APOE protein x protein interaction***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d4b64874",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========================================================================================================================================================================================================"
     ]
    }
   ],
   "source": [
    "# real\n",
    "# get protein names\n",
    "prot_names <- names(data_qc)[32:ncol(data_qc)]  # skip all covariates (check first)\n",
    "# create empty output sumstats table\n",
    "prot_sumstats <- data.frame(PROTEIN=numeric(), \n",
    "                            BETA=numeric(), \n",
    "                            STDERR=numeric(), \n",
    "                            PVALUE=numeric(), \n",
    "                            PROTxAPOE=numeric(),\n",
    "                            PROTxAPOE_STDERR=numeric(),\n",
    "                            PROTxAPOE_PVALUE=numeric(),\n",
    "                            NCASES=numeric(), \n",
    "                            NCONTROLS=numeric()\n",
    "                           )\n",
    "# progress bar\n",
    "pb = txtProgressBar(min = 0, max = length(prot_names), initial = 0) \n",
    "\n",
    "# run model: \"incident delirium ~ covariates + protein\" for each protein in prot_names\n",
    "for (i in 1:length(prot_names)) {\n",
    "#for (i in 1:2) {\n",
    "    #print(paste0(\"analyzing prlibrary(rhdf5)otein \", prot_names[i], \" ...\"))\n",
    "    ## skip APOE gene itself\n",
    "    if (prot_names[i]==\"APOE\") next\n",
    "    ## create formula string\n",
    "    formula_i  <- as.formula(\n",
    "        paste(\n",
    "            \"delirium_incident ~ sex + age_at_collection + BMI + APOE*\", prot_names[i]))\n",
    "    ## logistic reg model: delirium ~ coviates + protein_i\n",
    "    mod_i   <- glm(formula = formula_i, data = data_qc, family=\"binomial\",) %>% suppressWarnings()\n",
    "    ## extract coefficients: beta; std error; p-value\n",
    "    coefs_i <- summary(mod_i)$coefficients[prot_names[i],c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")] \n",
    "    ## append coefs in output sumstats table\n",
    "    prot_sumstats[i,c(\"BETA\",\"STDERR\",\"PVALUE\")] <- coefs_i\n",
    "    ## append protein_i name in output sumstats table\n",
    "    prot_sumstats[i,\"PROTEIN\"] <- prot_names[i]\n",
    "    ## append APOE interaction sumstats\n",
    "    prot_x_APOE_i       <- paste0(\"APOE:\", prot_names[i])\n",
    "    prot_x_APOE_coefs_i <- summary(mod_i)$coefficients[prot_x_APOE_i,c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")]\n",
    "    prot_sumstats[i,c(\"PROTxAPOE\",\"PROTxAPOE_STDERR\",\"PROTxAPOE_PVALUE\")] <- prot_x_APOE_coefs_i\n",
    "    ## append protein_i no. of cases + controls (using del_per_prot_train table)\n",
    "    ncases    <- del_per_prot[2,prot_names[i]] # 2nd row: cases\n",
    "    ncontrols <- del_per_prot[1,prot_names[i]] # 1st row: cases\n",
    "    prot_sumstats[i,c(\"NCASES\",\"NCONTROLS\")] <- c(ncases, ncontrols)\n",
    "    #print(\"Done\")\n",
    "    setTxtProgressBar(pb,i)\n",
    "    #cat(paste0(\"\\rAnalysed \", i, \" out of \",  length(prot_names), \" proteins (\", round((i / length(prot_names)*100),2), \"%) ...\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "669583e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_sumstats_annot <- inner_join(x=prot_sumstats, y=assay, by=c(\"PROTEIN\"=\"Assay\")) \n",
    "# make UniProt var to merge \n",
    "prot_sumstats_annot <- prot_sumstats_annot %>% mutate(UniProt2 = gsub(\"-.*\",\"\", UniProt)) %>% mutate(UniProt2 = gsub(\"_.*\",\"\", UniProt2))\n",
    "\n",
    "prot_sumstats_annot_pos <- \n",
    "left_join(x=prot_sumstats_annot, y=pos[,-c(\"Assay_Target\",\"Gene_symbol\")], by=c(\"UniProt2\"=\"UniProt\")) %>% \n",
    "    mutate(Gene_CHROM = gsub(\"X\",\"23\",Gene_CHROM)) %>%\n",
    "    mutate(Gene_CHROM = as.numeric(Gene_CHROM)) %>%\n",
    "    mutate(Gene_start = as.numeric(Gene_start)) %>%\n",
    "    mutate(Gene_end = as.numeric(Gene_end)) %>%\n",
    "    arrange(Gene_CHROM, Gene_start)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e90dc048",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>2918</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 2918\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 2918\n",
       "2. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 2918   15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROTEIN</th><th scope=col>BETA</th><th scope=col>STDERR</th><th scope=col>PVALUE</th><th scope=col>PROTxAPOE</th><th scope=col>PROTxAPOE_STDERR</th><th scope=col>PROTxAPOE_PVALUE</th><th scope=col>NCASES</th><th scope=col>NCONTROLS</th><th scope=col>UniProt</th><th scope=col>Panel</th><th scope=col>UniProt2</th><th scope=col>Gene_CHROM</th><th scope=col>Gene_start</th><th scope=col>Gene_end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PIGR             </td><td> 0.27759302</td><td>0.04709221</td><td>3.754331e-09</td><td> 0.1652197</td><td>0.04063106</td><td>4.775737e-05</td><td>523</td><td>31407</td><td>P01833</td><td>Neurology         </td><td>P01833</td><td> 1</td><td>206928522</td><td>206946466</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CXADR            </td><td> 0.13578651</td><td>0.04537316</td><td>2.765561e-03</td><td> 0.1470047</td><td>0.04102426</td><td>3.392034e-04</td><td>533</td><td>31945</td><td>P78310</td><td>Inflammation      </td><td>P78310</td><td>21</td><td> 17513043</td><td> 17593579</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>GGT1             </td><td> 0.16073518</td><td>0.04945694</td><td>1.154039e-03</td><td> 0.1395505</td><td>0.04017910</td><td>5.142720e-04</td><td>531</td><td>31833</td><td>P19440</td><td>Neurology         </td><td>P19440</td><td>22</td><td> 24594811</td><td> 24629005</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>FDX1             </td><td> 0.04253069</td><td>0.04564553</td><td>3.514605e-01</td><td> 0.1500576</td><td>0.04409584</td><td>6.665338e-04</td><td>541</td><td>32111</td><td>P10109</td><td>Cardiometabolic II</td><td>P10109</td><td>11</td><td>110429948</td><td>110464884</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>OFD1             </td><td>-0.01925345</td><td>0.04425663</td><td>6.635326e-01</td><td>-0.1430894</td><td>0.04334138</td><td>9.618668e-04</td><td>541</td><td>32111</td><td>O75665</td><td>Oncology II       </td><td>O75665</td><td>23</td><td> 13734743</td><td> 13777955</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>DEFB103A_DEFB103B</td><td>-0.04728862</td><td>0.04489113</td><td>2.921546e-01</td><td>-0.1301655</td><td>0.04009102</td><td>1.167312e-03</td><td>526</td><td>31865</td><td>P81534</td><td>Inflammation II   </td><td>P81534</td><td> 8</td><td>  7428888</td><td>  7430348</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & PROTEIN & BETA & STDERR & PVALUE & PROTxAPOE & PROTxAPOE\\_STDERR & PROTxAPOE\\_PVALUE & NCASES & NCONTROLS & UniProt & Panel & UniProt2 & Gene\\_CHROM & Gene\\_start & Gene\\_end\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & PIGR              &  0.27759302 & 0.04709221 & 3.754331e-09 &  0.1652197 & 0.04063106 & 4.775737e-05 & 523 & 31407 & P01833 & Neurology          & P01833 &  1 & 206928522 & 206946466\\\\\n",
       "\t2 & CXADR             &  0.13578651 & 0.04537316 & 2.765561e-03 &  0.1470047 & 0.04102426 & 3.392034e-04 & 533 & 31945 & P78310 & Inflammation       & P78310 & 21 &  17513043 &  17593579\\\\\n",
       "\t3 & GGT1              &  0.16073518 & 0.04945694 & 1.154039e-03 &  0.1395505 & 0.04017910 & 5.142720e-04 & 531 & 31833 & P19440 & Neurology          & P19440 & 22 &  24594811 &  24629005\\\\\n",
       "\t4 & FDX1              &  0.04253069 & 0.04564553 & 3.514605e-01 &  0.1500576 & 0.04409584 & 6.665338e-04 & 541 & 32111 & P10109 & Cardiometabolic II & P10109 & 11 & 110429948 & 110464884\\\\\n",
       "\t5 & OFD1              & -0.01925345 & 0.04425663 & 6.635326e-01 & -0.1430894 & 0.04334138 & 9.618668e-04 & 541 & 32111 & O75665 & Oncology II        & O75665 & 23 &  13734743 &  13777955\\\\\n",
       "\t6 & DEFB103A\\_DEFB103B & -0.04728862 & 0.04489113 & 2.921546e-01 & -0.1301655 & 0.04009102 & 1.167312e-03 & 526 & 31865 & P81534 & Inflammation II    & P81534 &  8 &   7428888 &   7430348\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 15\n",
       "\n",
       "| <!--/--> | PROTEIN &lt;chr&gt; | BETA &lt;dbl&gt; | STDERR &lt;dbl&gt; | PVALUE &lt;dbl&gt; | PROTxAPOE &lt;dbl&gt; | PROTxAPOE_STDERR &lt;dbl&gt; | PROTxAPOE_PVALUE &lt;dbl&gt; | NCASES &lt;dbl&gt; | NCONTROLS &lt;dbl&gt; | UniProt &lt;chr&gt; | Panel &lt;chr&gt; | UniProt2 &lt;chr&gt; | Gene_CHROM &lt;dbl&gt; | Gene_start &lt;dbl&gt; | Gene_end &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | PIGR              |  0.27759302 | 0.04709221 | 3.754331e-09 |  0.1652197 | 0.04063106 | 4.775737e-05 | 523 | 31407 | P01833 | Neurology          | P01833 |  1 | 206928522 | 206946466 |\n",
       "| 2 | CXADR             |  0.13578651 | 0.04537316 | 2.765561e-03 |  0.1470047 | 0.04102426 | 3.392034e-04 | 533 | 31945 | P78310 | Inflammation       | P78310 | 21 |  17513043 |  17593579 |\n",
       "| 3 | GGT1              |  0.16073518 | 0.04945694 | 1.154039e-03 |  0.1395505 | 0.04017910 | 5.142720e-04 | 531 | 31833 | P19440 | Neurology          | P19440 | 22 |  24594811 |  24629005 |\n",
       "| 4 | FDX1              |  0.04253069 | 0.04564553 | 3.514605e-01 |  0.1500576 | 0.04409584 | 6.665338e-04 | 541 | 32111 | P10109 | Cardiometabolic II | P10109 | 11 | 110429948 | 110464884 |\n",
       "| 5 | OFD1              | -0.01925345 | 0.04425663 | 6.635326e-01 | -0.1430894 | 0.04334138 | 9.618668e-04 | 541 | 32111 | O75665 | Oncology II        | O75665 | 23 |  13734743 |  13777955 |\n",
       "| 6 | DEFB103A_DEFB103B | -0.04728862 | 0.04489113 | 2.921546e-01 | -0.1301655 | 0.04009102 | 1.167312e-03 | 526 | 31865 | P81534 | Inflammation II    | P81534 |  8 |   7428888 |   7430348 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROTEIN           BETA        STDERR     PVALUE       PROTxAPOE  PROTxAPOE_STDERR PROTxAPOE_PVALUE NCASES NCONTROLS UniProt Panel              UniProt2 Gene_CHROM Gene_start Gene_end \n",
       "1 PIGR               0.27759302 0.04709221 3.754331e-09  0.1652197 0.04063106       4.775737e-05     523    31407     P01833  Neurology          P01833    1         206928522  206946466\n",
       "2 CXADR              0.13578651 0.04537316 2.765561e-03  0.1470047 0.04102426       3.392034e-04     533    31945     P78310  Inflammation       P78310   21          17513043   17593579\n",
       "3 GGT1               0.16073518 0.04945694 1.154039e-03  0.1395505 0.04017910       5.142720e-04     531    31833     P19440  Neurology          P19440   22          24594811   24629005\n",
       "4 FDX1               0.04253069 0.04564553 3.514605e-01  0.1500576 0.04409584       6.665338e-04     541    32111     P10109  Cardiometabolic II P10109   11         110429948  110464884\n",
       "5 OFD1              -0.01925345 0.04425663 6.635326e-01 -0.1430894 0.04334138       9.618668e-04     541    32111     O75665  Oncology II        O75665   23          13734743   13777955\n",
       "6 DEFB103A_DEFB103B -0.04728862 0.04489113 2.921546e-01 -0.1301655 0.04009102       1.167312e-03     526    31865     P81534  Inflammation II    P81534    8           7428888    7430348"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# output \n",
    "dim(prot_sumstats_annot_pos)\n",
    "prot_sumstats_annot_pos %>% arrange(PROTxAPOE_PVALUE) %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dcc507af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write output \n",
    "write.table(prot_sumstats_annot_pos, \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/output_pwas_models12/del_model1_glm_int_fullset_APOEint_sumstats.txt\", sep = \",\", row.names = F, col.names = T, quote = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "405e3206-5d5b-437a-913a-cbf56e406155",
   "metadata": {},
   "source": [
    "#### ***delirium ~ age + sex + BMI + APOE-e4***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "475b7e70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========================================================================================================================================================================================================"
     ]
    }
   ],
   "source": [
    "# real\n",
    "# get protein names\n",
    "prot_names <- names(data_qc)[32:ncol(data_qc)]  # skip all covariates (check first)\n",
    "# create empty output sumstats table\n",
    "prot_sumstats <- data.frame(PROTEIN=numeric(), \n",
    "                            BETA=numeric(), \n",
    "                            STDERR=numeric(), \n",
    "                            PVALUE=numeric(), \n",
    "                            NCASES=numeric(), \n",
    "                            NCONTROLS=numeric()\n",
    "                           )\n",
    "# progress bar\n",
    "pb = txtProgressBar(min = 0, max = length(prot_names), initial = 0) \n",
    "\n",
    "# run model: \"incident delirium ~ covariates + protein\" for each protein in prot_names\n",
    "for (i in 1:length(prot_names)) {\n",
    "#for (i in 1:2) {\n",
    "    #print(paste0(\"analyzing prlibrary(rhdf5)otein \", prot_names[i], \" ...\"))\n",
    "    ## create formula string\n",
    "    formula_i  <- as.formula(\n",
    "        paste(\n",
    "            \"delirium_incident ~ sex + age_at_collection + BMI + e4_count +\", prot_names[i]))\n",
    "    ## logistic reg model: delirium ~ coviates + protein_i\n",
    "    mod_i   <- glm(formula = formula_i, data = data_qc, family=\"binomial\",) %>% suppressWarnings()\n",
    "    ## extract coefficients: beta; std error; p-value\n",
    "    coefs_i <- summary(mod_i)$coefficients[prot_names[i],c(\"Estimate\",\"Std. Error\",\"Pr(>|z|)\")] \n",
    "    ## append coefs in output sumstats table\n",
    "    prot_sumstats[i,c(\"BETA\",\"STDERR\",\"PVALUE\")] <- coefs_i\n",
    "    ## append protein_i name in output sumstats table\n",
    "    prot_sumstats[i,\"PROTEIN\"] <- prot_names[i]\n",
    "    ## append protein_i no. of cases + controls (using del_per_prot_train table)\n",
    "    ncases    <- del_per_prot[2,prot_names[i]] # 2nd row: cases\n",
    "    ncontrols <- del_per_prot[1,prot_names[i]] # 1st row: cases\n",
    "    prot_sumstats[i,c(\"NCASES\",\"NCONTROLS\")] <- c(ncases, ncontrols)\n",
    "    #print(\"Done\")\n",
    "    setTxtProgressBar(pb,i)\n",
    "    #cat(paste0(\"\\rAnalysed \", i, \" out of \",  length(prot_names), \" proteins (\", round((i / length(prot_names)*100),2), \"%) ...\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4068e42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_sumstats_annot <- inner_join(x=prot_sumstats, y=assay, by=c(\"PROTEIN\"=\"Assay\")) \n",
    "# make UniProt var to merge \n",
    "prot_sumstats_annot <- prot_sumstats_annot %>% mutate(UniProt2 = gsub(\"-.*\",\"\", UniProt)) %>% mutate(UniProt2 = gsub(\"_.*\",\"\", UniProt2))\n",
    "\n",
    "prot_sumstats_annot_pos <- \n",
    "left_join(x=prot_sumstats_annot, y=pos[,-c(\"Assay_Target\",\"Gene_symbol\")], by=c(\"UniProt2\"=\"UniProt\")) %>% \n",
    "    mutate(Gene_CHROM = gsub(\"X\",\"23\",Gene_CHROM)) %>%\n",
    "    mutate(Gene_CHROM = as.numeric(Gene_CHROM)) %>%\n",
    "    mutate(Gene_start = as.numeric(Gene_start)) %>%\n",
    "    mutate(Gene_end = as.numeric(Gene_end)) %>%\n",
    "    arrange(Gene_CHROM, Gene_start)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a340d5ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>111</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 111\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 111\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 111  12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROTEIN</th><th scope=col>BETA</th><th scope=col>STDERR</th><th scope=col>PVALUE</th><th scope=col>NCASES</th><th scope=col>NCONTROLS</th><th scope=col>UniProt</th><th scope=col>Panel</th><th scope=col>UniProt2</th><th scope=col>Gene_CHROM</th><th scope=col>Gene_start</th><th scope=col>Gene_end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>GDF15</td><td>0.4492583</td><td>0.05525540</td><td>4.272476e-16</td><td>532</td><td>31884</td><td>Q99988</td><td>Cardiometabolic</td><td>Q99988</td><td>19</td><td>18374731</td><td>18389176</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>WFDC2</td><td>0.4106737</td><td>0.05112292</td><td>9.506805e-16</td><td>513</td><td>30513</td><td>Q14508</td><td>Oncology       </td><td>Q14508</td><td>20</td><td>45469753</td><td>45481532</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>AREG </td><td>0.3625946</td><td>0.04828552</td><td>5.940508e-14</td><td>533</td><td>31945</td><td>P15514</td><td>Oncology       </td><td>P15514</td><td> 4</td><td>74445136</td><td>74455005</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>IL6  </td><td>0.3385499</td><td>0.04862260</td><td>3.335561e-12</td><td>533</td><td>31945</td><td>P05231</td><td>Oncology       </td><td>P05231</td><td> 7</td><td>22725884</td><td>22732002</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PLAUR</td><td>0.3341755</td><td>0.04851640</td><td>5.662662e-12</td><td>520</td><td>31512</td><td>Q03405</td><td>Inflammation   </td><td>Q03405</td><td>19</td><td>43646095</td><td>43670547</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>TFF3 </td><td>0.3210395</td><td>0.04922193</td><td>6.924372e-11</td><td>527</td><td>31636</td><td>Q07654</td><td>Cardiometabolic</td><td>Q07654</td><td>21</td><td>42311667</td><td>42315409</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & PROTEIN & BETA & STDERR & PVALUE & NCASES & NCONTROLS & UniProt & Panel & UniProt2 & Gene\\_CHROM & Gene\\_start & Gene\\_end\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & GDF15 & 0.4492583 & 0.05525540 & 4.272476e-16 & 532 & 31884 & Q99988 & Cardiometabolic & Q99988 & 19 & 18374731 & 18389176\\\\\n",
       "\t2 & WFDC2 & 0.4106737 & 0.05112292 & 9.506805e-16 & 513 & 30513 & Q14508 & Oncology        & Q14508 & 20 & 45469753 & 45481532\\\\\n",
       "\t3 & AREG  & 0.3625946 & 0.04828552 & 5.940508e-14 & 533 & 31945 & P15514 & Oncology        & P15514 &  4 & 74445136 & 74455005\\\\\n",
       "\t4 & IL6   & 0.3385499 & 0.04862260 & 3.335561e-12 & 533 & 31945 & P05231 & Oncology        & P05231 &  7 & 22725884 & 22732002\\\\\n",
       "\t5 & PLAUR & 0.3341755 & 0.04851640 & 5.662662e-12 & 520 & 31512 & Q03405 & Inflammation    & Q03405 & 19 & 43646095 & 43670547\\\\\n",
       "\t6 & TFF3  & 0.3210395 & 0.04922193 & 6.924372e-11 & 527 & 31636 & Q07654 & Cardiometabolic & Q07654 & 21 & 42311667 & 42315409\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | PROTEIN &lt;chr&gt; | BETA &lt;dbl&gt; | STDERR &lt;dbl&gt; | PVALUE &lt;dbl&gt; | NCASES &lt;dbl&gt; | NCONTROLS &lt;dbl&gt; | UniProt &lt;chr&gt; | Panel &lt;chr&gt; | UniProt2 &lt;chr&gt; | Gene_CHROM &lt;dbl&gt; | Gene_start &lt;dbl&gt; | Gene_end &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | GDF15 | 0.4492583 | 0.05525540 | 4.272476e-16 | 532 | 31884 | Q99988 | Cardiometabolic | Q99988 | 19 | 18374731 | 18389176 |\n",
       "| 2 | WFDC2 | 0.4106737 | 0.05112292 | 9.506805e-16 | 513 | 30513 | Q14508 | Oncology        | Q14508 | 20 | 45469753 | 45481532 |\n",
       "| 3 | AREG  | 0.3625946 | 0.04828552 | 5.940508e-14 | 533 | 31945 | P15514 | Oncology        | P15514 |  4 | 74445136 | 74455005 |\n",
       "| 4 | IL6   | 0.3385499 | 0.04862260 | 3.335561e-12 | 533 | 31945 | P05231 | Oncology        | P05231 |  7 | 22725884 | 22732002 |\n",
       "| 5 | PLAUR | 0.3341755 | 0.04851640 | 5.662662e-12 | 520 | 31512 | Q03405 | Inflammation    | Q03405 | 19 | 43646095 | 43670547 |\n",
       "| 6 | TFF3  | 0.3210395 | 0.04922193 | 6.924372e-11 | 527 | 31636 | Q07654 | Cardiometabolic | Q07654 | 21 | 42311667 | 42315409 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROTEIN BETA      STDERR     PVALUE       NCASES NCONTROLS UniProt Panel           UniProt2 Gene_CHROM Gene_start Gene_end\n",
       "1 GDF15   0.4492583 0.05525540 4.272476e-16 532    31884     Q99988  Cardiometabolic Q99988   19         18374731   18389176\n",
       "2 WFDC2   0.4106737 0.05112292 9.506805e-16 513    30513     Q14508  Oncology        Q14508   20         45469753   45481532\n",
       "3 AREG    0.3625946 0.04828552 5.940508e-14 533    31945     P15514  Oncology        P15514    4         74445136   74455005\n",
       "4 IL6     0.3385499 0.04862260 3.335561e-12 533    31945     P05231  Oncology        P05231    7         22725884   22732002\n",
       "5 PLAUR   0.3341755 0.04851640 5.662662e-12 520    31512     Q03405  Inflammation    Q03405   19         43646095   43670547\n",
       "6 TFF3    0.3210395 0.04922193 6.924372e-11 527    31636     Q07654  Cardiometabolic Q07654   21         42311667   42315409"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# output \n",
    "prot_sumstats_annot_pos %>% arrange(PVALUE) %>% filter(PVALUE < 0.05 / nrow(.)) %>% dim\n",
    "prot_sumstats_annot_pos %>% arrange(PVALUE) %>% filter(PVALUE < 0.05 / nrow(.)) %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e367ebe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write output \n",
    "write.table(prot_sumstats_annot_pos, \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/proteomics/delirium/output_pwas_models12/del_model1_glm_int_fullset_e4Adj_sumstats.txt\", sep = \",\", row.names = F, col.names = T, quote = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e934a40",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f249724",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
